refactor: remove SettingsGateway

kyranet opened the pull request:
Fixes #1279.

- [x] Fill metadata for all keys in GuildEntity.
- [x] Fill metadata for complex keys (updatable via dashboard, soon also by Sapphire's dynamic argument system).
- [x] Make system to validate data.
- [x] Remove everything SettingsGateway related.
- [x] Add extendables for easy control of (guild) settings.
- [x] Update Klasa fork to remove all traces of SettingsGateway.
- [x] Update Skyra to patched commit of fork.
- [x] Write migration for `TEXT[]` to `JSONB`.
- [x] Merge Klasa PR (https://github.com/skyra-project/klasa/pull/3) then change Klasa dependency back to settings branch
- [x] Merge skyra.pw PR (https://github.com/skyra-project/skyra.pw/pull/198)
- [x] Reset depfu to target `main` branch for both this repo and skyra.pw
- [x] Test everything.
- [x] $$Profit$$

Check to do before merging:
- [x] formatting all
- [x] `readSettings` using array where possible
- [x] using `settings.getLanguage()` where possible
- [x] using `GuildMessage` when `runIn` is `['text']`
- [x] removing non-null assertion in `message.guild!.` when `runIn` is `['text']`
- [x] replacing `const { ... } = message.guild.readSettings((settings) => ({ ... }));` with `const [ ... ] = message.guild.readSettings((settings) => [ ... ])`
- [x] replacing `const { ... } = message.guild.writeSettings((settings) => ({ ... }));` with `const [ ... ] = message.guild.writeSettings((settings) => [ ... ])`
- [x] replacing `settings.key` with `settings[GuildSettings.PathToKey]`
- [x] temporarily enable `@typescript-eslint/no-unnecessary-type-assertion` as `"@typescript-eslint/no-unnecessary-type-assertion": "warn"` to check for this rule

kyranet committed:
refactor: remove SettingsGateway

kyranet committed:
refactor: remove SettingsGateway

Soumil-07 committed:
refactor: some more keys

Soumil-07 committed:
refactor: some more keys

kyranet committed:
refactor: folder re-organization

Soumil-07 committed:
refactor: finish sg keys

favna commented on src/tsconfig.json:
```diff
@@ -5,7 +5,6 @@
 		"rootDir": ".",
 		"baseUrl": ".",
 		"paths": {
-			"@orm/*": ["lib/orm/*"],
```
why? Why not `@database`?

favna commented on tests/tsconfig.json:
```diff
@@ -7,7 +7,6 @@
 		"incremental": false,
 		"baseUrl": ".",
 		"paths": {
-			"@orm/*": ["../src/lib/orm/*"],
```
why? Why not `@database`?

favna commented on src/lib/types/namespaces/languages/Settings.ts:
```diff
@@ -20,21 +20,23 @@ export const ChannelsIgnoreMessageDelete = T<string>('settingsChannelsIgnoreMess
+export const EventsBanRemove = T<string>('settingsEventsBanremove');
+export const EventsMemberAdd = T<string>('settingsEventsMemberadd');
+export const EventsMemberNameUpdate = T<string>('settingsEventsMembernameupdate');
+// TODO(kyranet): this key is missing in i18n.
```
TODO

favna commented on src/lib/database/settings/ConfigurableKey.ts:
```diff
@@ -0,0 +1,157 @@
+import type { CustomGet } from '@lib/types/Shared';
```
This file is mostly duplicated in `src/lib/util/Database/ConfigurableKey.ts`

favna commented on src/lib/database/settings/ConfigurableKey.ts:
```diff
@@ -0,0 +1,157 @@
+	/**
+	 * The type of the value this property accepts.
+	 */
+	type: string;
```
This needs to get strongly typed, i.e. through a const enum

kyranet committed:
chore: minor updates, make Skyra run

favna commented on src/lib/database/index.ts:
```diff
@@ -0,0 +1,25 @@
+export * from './entities/BannerEntity';
```
Why? Why not '@database'?

favna commented on src/lib/database/settings/ConfigurableKey.ts:
```diff
@@ -0,0 +1,117 @@
+}
+
+type ConfigurableKeyOptions = Omit<ConfigurableKeyValueOptions, 'name' | 'target' | 'property' | 'type' | 'maximum' | 'minimum' | 'array'> &
+	Partial<Pick<ConfigurableKeyValueOptions, 'name' | 'type' | 'maximum' | 'minimum' | 'array'>>;
```
This is partially duplicated in src/lib/database/settings/ConfigurableKeyValue.ts

favna commented on src/lib/database/settings/ConfigurableKeyValue.ts:
```diff
@@ -0,0 +1,73 @@
+	}
+}
+
+export type ConfigurableKeyValueOptions = Pick<
```
This is partially duplicated in src/lib/database/settings/ConfigurableKey.ts

favna commented on src/lib/util/Database/ConfigurableKey.ts:
```diff
@@ -0,0 +1,157 @@
+	/**
+	 * The type of the value this property accepts.
+	 */
+	type: string;
```
This needs to get strongly typed, i.e. through a const enum

favna commented on src/lib/util/Database/ConfigurableKey.ts:
```diff
@@ -0,0 +1,157 @@
+import type { CustomGet } from '@lib/types/Shared';
```
This file is mostly duplicated in `src/lib/database/settings/ConfigurableKey.ts`

kyranet committed:
fix: resolved build error

kyranet committed:
chore: first start at removing SG

kyranet committed:
fix: resolved a lot of errors

kyranet committed:
feat: easier internals

kyranet committed:
refactor: remove SG's schema

kyranet committed:
fix: resolved a few bugs in GuildEntity

Soumil-07 committed:
refactor: fix tools and twitch commands

Soumil-07 committed:
refactor: somewhat fixed serializers

kyranet committed:
chore: progress

kyranet committed:
chore: more progress

favna committed:
fix: remove a bunch more errors (1444 -> 1315)

Co-authored-by: Antonio Román <kyradiscord@gmail.com>

favna committed:
fix: fixed a bunch of errors

Co-authored-by: QuantumlyTangled <ethoarx@gmail.com>
Co-authored-by: Antonio Román <kyradiscord@gmail.com>
Co-authored-by: Charalampos Fanoulis <charalampos.fanoulis@gmail.com>
Co-authored-by: Rexogamer <Rexogamer@users.noreply.github.com>

favna committed:
fix: more fixes

Co-authored-by: Antonio Román <kyradiscord@gmail.com>
Co-authored-by: Charalampos Fanoulis <charalampos.fanoulis@gmail.com>

favna committed:
refactor: code cleanup

kyranet committed:
chore: update Skyra to PR

favna committed:
build: update lockfile

favna committed:
chore: nitpickery extremus
aka switch to sapphire tsconfig and eslint config

favna committed:
chore: bump sapphire packages

favna committed:
build: fix pm2 ecosystem file format

favna committed:
fix: bunch o' fixes in 'ere matey

Co-authored-by: Antonio Román <kyradiscord@gmail.com>

depfu[bot] committed:
Update discord-api-types to version 0.8.0 (#1368)

Co-authored-by: depfu[bot] <23717796+depfu[bot]@users.noreply.github.com>

kyranet committed:
chore: many enhancements

kyranet committed:
chore: fixed more stuff

kyranet committed:
chore: more fixes

kyranet committed:
chore: more fixes

kyranet committed:
chore: more fixes

kyranet committed:
chore: more bugfixes

favna committed:
chore: swap out klasa packages for sapphire packages

kyranet committed:
chore: remove fs-nextra

kyranet committed:
chore: fixed more errors, minor file re-organization

kyranet committed:
chore: move DbSet, removed friendlyDuration tests

kyranet committed:
fix: re-order imports

favna committed:
ci: use stable version of artifact action

kyranet committed:
fix: automatically expire keys from Redis, fixed scan bug

favna committed:
chore: moaaarr fixes

Co-authored-by: Antonio Román <kyradiscord@gmail.com>

vladfrangu commented on package.json:
```diff
@@ -12,7 +12,7 @@
 		"start": "node dist/Skyra.js",
 		"lint": "eslint --fix src tests --ext ts",
-		"format": "prettier --write --loglevel=warn \"src/**/*.ts\"",
+		"format": "prettier --write --loglevel=warn \"{src,tests}/**/*.ts\"",
```
Why aren't you using eslint's fix? (since prettier rules are included)

kyranet replied:
Because we already have a lint script above, prettier does its job a lot faster.

kyranet committed:
fix: resolved ALL BUILD ERRORS

HELL YEAH!

vladfrangu commented on src/arguments/song.ts:
```diff
@@ -30,11 +30,11 @@ export default class extends Argument {
 		const { id } = message.author;
 		const entries = count(tracks.values(), (track) => track.author === id);
-		const maximum = message.guild!.settings.get(GuildSettings.Music.MaximumEntriesPerUser);
+		const maximum = await message.guild.readSettings((guild) => guild.musicMaximumEntriesPerUser);
```
From what I saw in diff files, you can just fetch by key?

kyranet committed:
fix: resolved tests error

vladfrangu commented on src/commands/Animal/catfact.ts:
```diff
@@ -22,7 +22,7 @@ export default class extends SkyraCommand {
 			new MessageEmbed()
 				.setColor(await DbSet.fetchColor(message))
-				.setTitle(message.language.get(LanguageKeys.Commands.Animal.CatfactTitle))
+				.setTitle(message.fetchLocale(LanguageKeys.Commands.Animal.CatfactTitle))
```
missing await?

vladfrangu commented on src/commands/Anime/waifu.ts:
```diff
@@ -27,7 +27,7 @@ export default class extends SkyraCommand {
 				.setColor(await DbSet.fetchColor(message))
 				.setImage(url)
-				.setFooter(message.language.get(LanguageKeys.Commands.Anime.WaifuFooter))
+				.setFooter(message.fetchLocale(LanguageKeys.Commands.Anime.WaifuFooter))
```
Missing await?

vladfrangu commented on src/commands/Announcement/announcement.ts:
```diff
@@ -45,23 +48,28 @@ export default class extends SkyraCommand {
 
-	private ask(message: KlasaMessage, header: string, announcement: string) {
+	private async ask(message: GuildMessage, header: string, announcement: string) {
+		const language = await message.fetchLanguage();
```
Can't you pass in the language from above?

vladfrangu commented on src/commands/Announcement/announcement.ts:
```diff
@@ -70,11 +78,11 @@ export default class extends SkyraCommand {
+								? language.get(LanguageKeys.Commands.Announcement.AnnouncementEmbedMentionsWithMentions, {
 										header,
-										mentions: message.language.list(mentions, message.language.get(LanguageKeys.Globals.And))
+										mentions: language.list(mentions, await message.fetchLocale(LanguageKeys.Globals.And))
```
Why are you fetchLocale here when you have language?
```suggestion
										mentions: language.list(mentions, language.get(LanguageKeys.Globals.And))
```

vladfrangu commented on src/commands/Fun/dice.ts:
```diff
@@ -31,30 +31,31 @@ export default class extends SkyraCommand {
+		const language = await message.fetchLanguage();
 		if (typeof pattern === 'number') {
-			if (!isNumber(pattern) || pattern <= 0) throw message.language.get(LanguageKeys.Resolvers.InvalidInt, { name: 'dice' });
+			if (!isNumber(pattern) || pattern <= 0) throw language.get(LanguageKeys.Resolvers.InvalidInt, { name: 'dice' });
```
nit: name isn't i18n'd

kyranet replied:
That's the command's name.

vladfrangu commented on src/commands/Fun/escaperope.ts:
```diff
@@ -16,12 +16,13 @@ import { KlasaMessage } from 'klasa';
 				.setColor(await DbSet.fetchColor(message))
 				.setImage(CdnUrls.EscapeRopeGif)
-				.setDescription(message.language.get(LanguageKeys.Commands.Fun.EscaperopeOutput, { user: message.author }))
+				.setDescription(language.get(LanguageKeys.Commands.Fun.EscaperopeOutput, { user: message.author }))
```
Don't you want toString()?

vladfrangu commented on src/commands/Fun/markov.ts:
```diff
@@ -38,37 +39,41 @@ export default class extends SkyraCommand {
 
-	private async retrieveMarkov(message: KlasaMessage, user: User | undefined, channel: TextChannel = message.channel as TextChannel) {
+	private async retrieveMarkov(message: GuildMessage, user: User | undefined, channel: TextChannel = message.channel as TextChannel) {
+		const language = await message.fetchLanguage();
```
re-use language from run block

vladfrangu commented on src/commands/Fun/rate.ts:
```diff
@@ -17,13 +17,14 @@ export default class extends SkyraCommand {
 		if (/^(you|yourself|skyra)$/i.test(user)) {
 			rate = 100;
-			[ratewaifu, user] = message.language.get(LanguageKeys.Commands.Fun.RateMyself);
+			[ratewaifu, user] = language.get(LanguageKeys.Commands.Fun.RateMyself);
```
q: does this actually return an array?

kyranet replied:
Yes, otherwise TS wouldn't accept this.

vladfrangu commented on src/commands/GameIntegration/brawlstars.ts:
```diff
@@ -110,8 +109,9 @@ export default class extends SkyraCommand {
-		const fields = message.language.get(LanguageKeys.Commands.GameIntegration.BrawlstarsPlayerEmbedFields);
+		const titles = await message.fetchLocale(LanguageKeys.Commands.GameIntegration.BrawlstarsPlayerEmbedTitles);
+		const fields = await message.fetchLocale(LanguageKeys.Commands.GameIntegration.BrawlstarsPlayerEmbedFields);
+		const languageString = (await message.fetchLanguage()).name;
```
Assuming `fetchLanguage` returns `Language`, change to 1 call to fetchLanguage and the rest accordingly

vladfrangu commented on src/commands/GameIntegration/brawlstars.ts:
```diff
@@ -167,12 +163,13 @@ export default class extends SkyraCommand {
-		const fields = message.language.get(LanguageKeys.Commands.GameIntegration.BrawlstarsClubEmbedFields);
+		const titles = await message.fetchLocale(LanguageKeys.Commands.GameIntegration.BrawlstarsClubEmbedTitles);
+		const fields = await message.fetchLocale(LanguageKeys.Commands.GameIntegration.BrawlstarsClubEmbedFields);
+		const languageString = (await message.fetchLanguage()).name;
```
Ditto~

vladfrangu commented on src/commands/GameIntegration/overwatch.ts:
```diff
@@ -41,31 +45,34 @@ export default class extends RichDisplayCommand {
-	private async buildDisplay(message: KlasaMessage, overwatchData: OverwatchDataSet, player: string, platform: PlatformUnion) {
-		let ratings = Array.from(
+	private async buildDisplay(message: GuildMessage, overwatchData: OverwatchDataSet, player: string, platform: PlatformUnion) {
+		const language = await message.fetchLanguage();
```
get language from run block

vladfrangu commented on src/commands/GameIntegration/overwatch.ts:
```diff
@@ -147,7 +159,7 @@ export default class extends RichDisplayCommand {
 
 		const timePlayedMilliseconds = Number(timePlayed.split(':')[0]) * Time.Hour + Number(timePlayed.split(':')[1]) * Time.Minute;
-		const statsData = message.language.get(LanguageKeys.Commands.GameIntegration.OverwatchEmbedDataStats, {
+		const statsData = await message.fetchLocale(LanguageKeys.Commands.GameIntegration.OverwatchEmbedDataStats, {
```
Can't you get language from above?

vladfrangu commented on src/commands/Games/coinflip.ts:
```diff
@@ -43,63 +64,45 @@ export default class extends SkyraCommand {
 
 	private async noGuess(message: KlasaMessage) {
 		const result = this.flipCoin();
+		const language = await message.fetchLanguage();
```
Fetch language at the top of run, pass it down

vladfrangu commented on src/commands/Games/coinflip.ts:
```diff
@@ -43,63 +64,45 @@ export default class extends SkyraCommand {
 	private async cashless(message: KlasaMessage, guess: CoinType) {
 		const result = this.flipCoin();
 		const won = result === guess;
+		const language = await message.fetchLanguage();
```
As above/below

vladfrangu commented on src/commands/Games/hungergames.ts:
```diff
@@ -38,13 +39,16 @@ export default class extends SkyraCommand {
-		if (filtered.size !== tributes.length) throw message.language.get(LanguageKeys.Commands.Games.GamesRepeat);
-		if (this.playing.has(message.channel.id)) throw message.language.get(LanguageKeys.Commands.Games.GamesProgress);
-		if (filtered.size < 4 || filtered.size > 48) throw message.language.get(LanguageKeys.Commands.Games.GamesTooManyOrFew, { min: 4, max: 48 });
+		const language = await message.fetchLanguage();
```
nit?: move this above, re-use in else if check

kyranet replied:
Not applicable, the above check also fetches prefix, and always throws.

vladfrangu commented on src/commands/Games/tictactoe.ts:
```diff
@@ -46,17 +46,18 @@ export default class extends SkyraCommand {
+		if (user.bot) throw await message.fetchLocale(LanguageKeys.Commands.Games.GamesBot);
+		if (user.id === message.author.id) throw await message.fetchLocale(LanguageKeys.Commands.Games.GamesSelf);
 
+		const language = await message.fetchLanguage();
```
fetch above, reuse in if checks

vladfrangu commented on src/commands/Games/trivia.ts:
```diff
@@ -50,20 +50,21 @@ export default class extends SkyraCommand {
 
 		this.#channels.add(message.channel.id);
 
+		const language = await message.fetchLanguage();
```
fetch above, re-use in if check

vladfrangu commented on src/commands/Games/trivia.ts:
```diff
@@ -50,20 +50,21 @@ export default class extends SkyraCommand {
+		const language = await message.fetchLanguage();
 		try {
-			await message.send(pickRandom(message.language.get(LanguageKeys.System.Loading)), []);
+			await message.send(pickRandom(language.get(LanguageKeys.System.Loading)), []);
```
whats that empty array doing there?

vladfrangu commented on src/commands/Management/Attachment Filter/manageAttachments.ts:
```diff
@@ -1,135 +1,146 @@
+				const value = await message.client.arguments.get('boolean')!.run(arg, possible, message);
+				return message.guild!.readSettings((settings) =>
+					value
+						? (settings[GuildSettings.Selfmod.Attachments.Action] & 0b0111) | 0b1000
```
magic numbers ;w;
Constant them up when?

kyranet replied:
When they formally become part of the rest of automod systems, which needs a migration and a lot moves, pending for another PR.

vladfrangu commented on src/commands/Management/Configuration/manageCommandAutodelete.ts:
```diff
@@ -26,62 +26,70 @@ import { KlasaMessage } from 'klasa';
 		}
-		if (!list.length) throw message.language.get(LanguageKeys.Commands.Management.ManageCommandAutoDeleteShowEmpty);
+		if (!list.length) throw language.get(LanguageKeys.Commands.Management.ManageCommandAutoDeleteShowEmpty);
 		return message.sendLocale(LanguageKeys.Commands.Management.ManageCommandAutoDeleteShow, [
```
Should move to msg.send me thinks

vladfrangu commented on src/commands/Management/Configuration/manageCommandAutodelete.ts:
```diff
@@ -26,62 +26,70 @@ import { KlasaMessage } from 'klasa';
+			await message.guild.writeSettings((settings) => {
+				const values = settings[GuildSettings.CommandAutodelete];
+				values.push(value);
+				return settings.getLanguage();
```
Returns are..useless..?

vladfrangu commented on src/commands/Management/Configuration/manageCommandAutodelete.ts:
```diff
@@ -26,62 +26,70 @@ import { KlasaMessage } from 'klasa';
+			await message.guild.writeSettings((settings) => {
+				const values = settings[GuildSettings.CommandAutodelete];
+				values.splice(index, 1);
+				return settings.getLanguage();
```
Ditto

vladfrangu commented on src/commands/Management/Configuration/manageCommandAutodelete.ts:
```diff
@@ -26,62 +26,70 @@ import { KlasaMessage } from 'klasa';
+		await message.guild.writeSettings((settings) => {
+			const values = settings[GuildSettings.CommandAutodelete];
+			values.length = 0;
+			return settings.getLanguage();
```
BOnk

vladfrangu commented on src/commands/Management/Configuration/setIgnoreChannels.ts:
```diff
@@ -21,12 +21,18 @@ export default class extends SkyraCommand {
+				ignoredChannels.splice(ignoredChannels.indexOf((channel as TextChannel).id), 1);
+			else ignoredChannels.push((channel as TextChannel).id);
+
+			return settings.getLanguage();
```
Uselesss return, if anything you should return newLength here

vladfrangu commented on src/commands/Management/Configuration/setImageLogs.ts:
```diff
@@ -1,33 +1,40 @@
-		else if (!isTextBasedChannel(channel)) throw message.language.get(LanguageKeys.Misc.ConfigurationTextChannelRequired);
+		const channelID = channel.id;
+
+		const language = await message.guild.writeSettings((settings) => {
```
You know I love you right?

I know you love me too!

Make this duplicated code..not be a thing by creating a base class that does all of this, and all you have to set is the appropriate keys

kyranet replied:
Placed a TODO, it's out of this PR's scope.

vladfrangu commented on src/commands/Management/Helpers/guildInfo.ts:
```diff
@@ -1,80 +1,80 @@
 		let cChannels = 0;
-		for (const channel of message.guild!.channels.cache.values()) {
+		for (const channel of message.guild.channels.cache.values()) {
 			if (channel.type === 'text') tChannels++;
```
This doesn't account for news channels

vladfrangu commented on src/commands/Management/Helpers/guildInfo.ts:
```diff
@@ -1,80 +1,80 @@
+		for (const channel of message.guild.channels.cache.values()) {
 			if (channel.type === 'text') tChannels++;
 			else if (channel.type === 'voice') vChannels++;
 			else cChannels++;
```
store channels will increase this

vladfrangu commented on src/commands/Management/Members/stickyRoles.ts:
```diff
@@ -20,56 +20,57 @@ import { KlasaMessage } from 'klasa';
-		const names = roles.map((role) => guildRoles.cache.get(role)!.name);
-		return message.sendLocale(LanguageKeys.Commands.Management.StickyRolesShowSingle, [
-			{
+		const language = await message.fetchLanguage();
```
Fetch, re-use in if check

vladfrangu commented on src/commands/Management/Members/stickyRoles.ts:
```diff
@@ -20,56 +20,57 @@ import { KlasaMessage } from 'klasa';
-			{
+		const language = await message.fetchLanguage();
+		const roles = message.guild.roles.cache;
+		const names = sticky.map((role) => roles.get(role)!.name);
```
Assuming sticky roles don't update when a role is removed, this CAN throw

kyranet replied:
~~Placed a TODO, it's the same code as before, and this needs a new event, out of this PR's scope.~~

Actually, it never can, `StickyRoles.fetch` removes deleted roles.

vladfrangu commented on src/commands/Management/Message Filters/filter.ts:
```diff
@@ -1,66 +1,91 @@
+		const language = await message.guild.writeSettings((settings) => {
+			const language = settings.getLanguage();
+
+			// Check if the word is already filtered:
```
```suggestion
			// Check if the word is not filtered:
```

vladfrangu commented on src/commands/Management/Message Filters/filter.ts:
```diff
@@ -1,66 +1,91 @@
+			const language = settings.getLanguage();
+
+			// Set an empty array:
+			settings[GuildSettings.Selfmod.Filter.Raw] = [];
```
```suggestion
			settings[GuildSettings.Selfmod.Filter.Raw].length = 0;
```

vladfrangu commented on src/commands/Management/permissionNodes.ts:
```diff
@@ -1,143 +1,155 @@
+		const { nodes, language } = await message.guild.readSettings((settings) => ({
+			nodes: settings[key],
+			language: settings.getLanguage()
+		}));
```
```suggestion
		const [ nodes, language ] = await message.guild.readSettings((settings) => [
			settings[key],
			settings.getLanguage()
		]);
```

vladfrangu commented on src/commands/Management/triggers.ts:
```diff
@@ -22,80 +21,87 @@ const REG_TYPE = /^(alias|reaction)$/i;
+			if (!arg) throw await message.fetchLocale(LanguageKeys.Commands.Management.TriggersNooutput);
+			if (type === 'reaction') {
+				const emoji = resolveEmoji(arg);
+				if (!emoji) throw await message.fetchLocale(LanguageKeys.Commands.Management.TriggersInvalidreaction);
```
In the fuck
Please fix the lowercase `r`

vladfrangu commented on src/commands/Management/triggers.ts:
```diff
@@ -22,80 +21,87 @@ const REG_TYPE = /^(alias|reaction)$/i;
+		'output',
+		async (arg, _, message, [action, type]) => {
+			if (action === 'show' || action === 'remove') return undefined;
+			if (!arg) throw await message.fetchLocale(LanguageKeys.Commands.Management.TriggersNooutput);
```
Fix `Nooutput` -> `NoOutput`

vladfrangu commented on src/commands/Management/triggers.ts:
```diff
@@ -22,80 +21,87 @@ const REG_TYPE = /^(alias|reaction)$/i;
-					throw message.language.get(LanguageKeys.Commands.Management.TriggersInvalidalias);
-				} else {
-					return null;
+			throw await message.fetchLocale(LanguageKeys.Commands.Management.TriggersNotype);
```
Fix `Notype` -> `NoType`

vladfrangu commented on src/commands/Moderation/Restriction/restrictAttachment.ts:
```diff
@@ -24,21 +24,24 @@ export default class extends ModerationCommand {
+		const { id, language } = await message.guild.readSettings((settings) => ({
+			id: settings[GuildSettings.Roles.RestrictedAttachment],
+			language: settings.getLanguage()
+		}));
```
```suggestion
		const [id, language] = await message.guild.readSettings((settings) => [
			settings[GuildSettings.Roles.RestrictedAttachment],
			settings.getLanguage()
		]);
```

vladfrangu commented on src/commands/Moderation/Restriction/restrictEmbed.ts:
```diff
@@ -24,19 +24,22 @@ export default class extends ModerationCommand {
+		const { id, language } = await message.guild.readSettings((settings) => ({
+			id: settings[GuildSettings.Roles.RestrictedEmbed],
+			language: settings.getLanguage()
+		}));
```
```suggestion
		const [id, language] = await message.guild.readSettings((settings) => [
			settings[GuildSettings.Roles.RestrictedEmbed],
			settings.getLanguage()
		]);
```

vladfrangu commented on src/commands/Moderation/Restriction/restrictEmoji.ts:
```diff
@@ -24,19 +24,22 @@ export default class extends ModerationCommand {
+		const { id, language } = await message.guild.readSettings((settings) => ({
+			id: settings[GuildSettings.Roles.RestrictedEmoji],
+			language: settings.getLanguage()
+		}));
```
```suggestion
		const [id, language] = await message.guild.readSettings((settings) => [
			settings[GuildSettings.Roles.RestrictedEmoji],
			settings.getLanguage()
		]);
```

vladfrangu commented on src/commands/Moderation/Restriction/restrictReaction.ts:
```diff
@@ -24,19 +24,22 @@ export default class extends ModerationCommand {
+		const { id, language } = await message.guild.readSettings((settings) => ({
+			id: settings[GuildSettings.Roles.RestrictedReaction],
+			language: settings.getLanguage()
+		}));
```
```suggestion
		const [id, language] = await message.guild.readSettings((settings) => [
			settings[GuildSettings.Roles.RestrictedReaction],
			settings.getLanguage()
		]);
```

vladfrangu commented on src/commands/Moderation/Restriction/restrictVoice.ts:
```diff
@@ -21,22 +21,25 @@ export default class extends ModerationCommand {
+		const { id, language } = await message.guild.readSettings((settings) => ({
+			id: settings[GuildSettings.Roles.RestrictedVoice],
+			language: settings.getLanguage()
+		}));
```
```suggestion
		const [id, language] = await message.guild.readSettings((settings) => [
			settings[GuildSettings.Roles.RestrictedVoice],
			settings.getLanguage()
		]);
```

vladfrangu commented on src/commands/Moderation/Restriction/unrestrictAttachment.ts:
```diff
@@ -14,13 +14,18 @@ import { KlasaMessage } from 'klasa';
+			id: settings[this.kPath],
+			prefix: settings[GuildSettings.Prefix],
+			language: settings.getLanguage()
+		}));
```
```suggestion
		const [id, prefix, language] = await message.guild.readSettings((settings) => [
			settings[this.kPath],
			settings[GuildSettings.Prefix],
			settings.getLanguage()
		]);
```

vladfrangu commented on src/commands/Moderation/Restriction/unrestrictEmbed.ts:
```diff
@@ -14,13 +14,18 @@ import { KlasaMessage } from 'klasa';
+			id: settings[this.kPath],
+			prefix: settings[GuildSettings.Prefix],
+			language: settings.getLanguage()
+		}));
```
```suggestion
		const [id, prefix, language] = await message.guild.readSettings((settings) => [
			id: settings[this.kPath],
			prefix: settings[GuildSettings.Prefix],
			language: settings.getLanguage()
		]);
```

vladfrangu commented on src/commands/Moderation/Restriction/unrestrictEmoji.ts:
```diff
@@ -14,13 +14,19 @@ import { KlasaMessage } from 'klasa';
+			id: settings[this.kPath],
+			prefix: settings[GuildSettings.Prefix],
+			language: settings.getLanguage()
+		}));
```
```suggestion
		const [id, prefix, language] = await message.guild.readSettings((settings) => [
			settings[this.kPath],
			settings[GuildSettings.Prefix],
			settings.getLanguage()
		]);
```

vladfrangu commented on src/commands/Moderation/Restriction/unrestrictReaction.ts:
```diff
@@ -14,13 +14,18 @@ import { KlasaMessage } from 'klasa';
+			id: settings[this.kPath],
+			prefix: settings[GuildSettings.Prefix],
+			language: settings.getLanguage()
+		}));
```
```suggestion
		const [id, prefix, language] = await message.guild.readSettings((settings) => [
			settings[this.kPath],
			settings[GuildSettings.Prefix],
			settings.getLanguage()
		]);
```

vladfrangu commented on src/commands/Moderation/Restriction/unrestrictVoice.ts:
```diff
@@ -14,13 +14,19 @@ import { KlasaMessage } from 'klasa';
+			id: settings[this.kPath],
+			prefix: settings[GuildSettings.Prefix],
+			language: settings.getLanguage()
+		}));
```
```suggestion
		const [id, prefix, language] = await message.guild.readSettings((settings) => [
			settings[this.kPath],
			settings[GuildSettings.Prefix],
			settings.getLanguage()
		]);
```

vladfrangu commented on src/commands/Moderation/Utilities/permissions.ts:
```diff
@@ -1,46 +1,46 @@
 
 		const { permissions } = member;
 		const list = [ZeroWidthSpace];
+		const language = await message.fetchLanguage();
```
Fetch first, skip async in .catch

vladfrangu commented on src/commands/Moderation/softban.ts:
```diff
@@ -15,22 +15,21 @@ import { KlasaMessage } from 'klasa';
-			: null;
+	public async prehandle(...[message]: ArgumentTypes<ModerationCommand['prehandle']>) {
+		const [banAdd, banRemove] = await message.guild.readSettings([GuildSettings.Events.BanAdd, GuildSettings.Events.BanRemove]);
+		return banAdd || banRemove ? { unlock: message.guild.moderation.createLock() } : null;
```
extra small q: will JS interpret this as `(banAdd || banRemove) ? ...` or `banAdd || (banRemove ? ...)`?

kyranet replied:
Former, `||` has higher preference than `?`.

vladfrangu commented on src/commands/Music/exportqueue.ts:
```diff
@@ -25,8 +25,9 @@ export default class extends SkyraCommand {
-			message.language.get(LanguageKeys.Commands.Music.ExportQueueSuccess, { guildName: message.guild!.name }),
+		return message.sendLocale(
+			LanguageKeys.Commands.Music.ExportQueueSuccess,
+			[{ guildName: message.guild!.name }],
```
Use GuildMessage type

vladfrangu commented on src/commands/Music/promote.ts:
```diff
@@ -25,13 +25,14 @@ export default class extends MusicCommand {
 		if (index >= length) {
-			throw message.language.get(LanguageKeys.Commands.Music.RemoveIndexOutOfBounds, {
-				songs: message.language.get(
+			const language = await message.fetchLanguage();
```
Might as well fetch on top and then throw from it

vladfrangu commented on src/commands/Social/Member Management/social.ts:
```diff
@@ -85,21 +88,20 @@ export default class extends SkyraCommand {
 		const variation = amount - oldValue;
 		if (variation === 0) return message.sendLocale(LanguageKeys.Commands.Social.SocialUnchanged, [{ user: user.username }]);
+
+		const language = await message.fetchLanguage();
```
fetch above, skip sendLocale

vladfrangu commented on src/commands/Social/Profile Management/banner.ts:
```diff
@@ -34,30 +33,28 @@ export default class extends SkyraCommand {
-			throw message.language.get(LanguageKeys.Commands.Social.BannerBought, {
-				prefix: message.guild!.settings.get(GuildSettings.Prefix),
+			throw language.get(LanguageKeys.Commands.Social.BannerBought, {
+				prefix: await message.guild.readSettings(GuildSettings.Prefix),
```
read prefix in the main call (promise.all)

vladfrangu commented on src/commands/Social/Profile Management/banner.ts:
```diff
@@ -74,15 +71,15 @@ export default class extends SkyraCommand {
-				throw message.language.get(LanguageKeys.Commands.Social.BannerUserlistEmpty, {
-					prefix: message.guild!.settings.get(GuildSettings.Prefix)
+				throw language.get(LanguageKeys.Commands.Social.BannerUserlistEmpty, {
+					prefix: await message.guild.readSettings(GuildSettings.Prefix)
```
ditto

vladfrangu commented on src/commands/Social/Profile Management/banner.ts:
```diff
@@ -91,15 +88,15 @@ export default class extends SkyraCommand {
-				throw message.language.get(LanguageKeys.Commands.Social.BannerUserlistEmpty, {
-					prefix: message.guild!.settings.get(GuildSettings.Prefix)
+				throw language.get(LanguageKeys.Commands.Social.BannerUserlistEmpty, {
+					prefix: await message.guild.readSettings(GuildSettings.Prefix)
```
ditto

vladfrangu commented on src/commands/Social/Profile Management/banner.ts:
```diff
@@ -147,16 +146,20 @@ export default class extends SkyraCommand {
+		const { prefix, language } = await message.guild.readSettings((settings) => ({
+			prefix: settings[GuildSettings.Prefix],
+			language: settings.getLanguage()
+		}));
```
```suggestion
		const [prefix, language] = await message.guild.readSettings((settings) => [
			settings[GuildSettings.Prefix],
			settings.getLanguage()
		]);
```

vladfrangu commented on src/commands/Social/autorole.ts:
```diff
@@ -1,102 +1,121 @@
+	public async show(message: GuildMessage) {
+		const output = await message.guild.writeSettings((settings) => {
+			const autoRoles = settings[GuildSettings.Roles.Auto];
+			if (!autoRoles.length) throw settings.getLanguage().get(LanguageKeys.Commands.Social.AutoRoleListEmpty);
```
Extract getLanguage to a variable

vladfrangu commented on src/commands/Social/balance.ts:
```diff
@@ -18,13 +18,14 @@ export default class extends SkyraCommand {
-			? message.sendLocale(LanguageKeys.Commands.Social.BalanceSelf, [{ amount: message.language.groupDigits(money) }])
-			: message.sendLocale(LanguageKeys.Commands.Social.Balance, [{ user: user.username, amount: message.language.groupDigits(money) }]);
+			? message.sendLocale(LanguageKeys.Commands.Social.BalanceSelf, [{ amount: language.groupDigits(money) }])
+			: message.sendLocale(LanguageKeys.Commands.Social.Balance, [{ user: user.username, amount: language.groupDigits(money) }]);
```
replace with .send(language.get)

vladfrangu commented on src/commands/Social/daily.ts:
```diff
@@ -42,11 +42,12 @@ export default class extends SkyraCommand {
 
 			// It's been 11-12 hours, ask for the user if they want to claim the grace period
-			const accepted = await message.ask(message.language.get(LanguageKeys.Commands.Social.DailyGrace, { remaining }).join('\n'));
+			const accepted = await message.ask((await message.fetchLocale(LanguageKeys.Commands.Social.DailyGrace, { remaining })).join('\n'));
```
language.get

vladfrangu commented on src/commands/Social/leaderboard.ts:
```diff
@@ -55,17 +55,18 @@ export default class extends SkyraCommand {
+		const language = await message.fetchLanguage();
 		if (promises.length) {
-			await message.send(pickRandom(message.language.get(LanguageKeys.System.Loading)), []);
+			await message.send(pickRandom(language.get(LanguageKeys.System.Loading)), []);
```
What's that [] doing?

vladfrangu commented on src/commands/Social/pay.ts:
```diff
@@ -21,26 +22,27 @@ export default class extends SkyraCommand {
-				throw message.language.get(LanguageKeys.Commands.Social.PayMissingMoney, { needed: money, has: currencyBeforePrompt });
+				throw language.get(LanguageKeys.Commands.Social.PayMissingMoney, { needed: money, has: currencyBeforePrompt });
 
 			if (!accepted) return this.denyPayment(message);
```
add language here, re-use in functions

vladfrangu commented on src/commands/Suggestions/suggest.ts:
```diff
@@ -86,25 +85,27 @@ export default class extends SkyraCommand {
 			return true;
 		}
 
+		const language = await message.fetchLanguage();
```
Fetch before manageable check

vladfrangu commented on src/commands/Tags/tag.ts:
```diff
@@ -26,90 +25,95 @@ import { CommandOptions, KlasaMessage } from 'klasa';
+		const language = await message.fetchLanguage();
+		await message.guild.writeSettings((settings) => {
+			const tagIndex = settings[GuildSettings.CustomCommands].findIndex((command) => command.id === id);
+			if (tagIndex === -1) throw language.get(LanguageKeys.Commands.Tags.TagNotexists, { tag: id });
```
Fix casing (`TagNotExists`)

vladfrangu commented on src/commands/Tools/Dictionary/urban.ts:
```diff
@@ -44,9 +44,11 @@ export default class extends RichDisplayCommand {
 				.setThumbnail('https://i.imgur.com/CcIZZsa.png')
 		).setFooterSuffix(' - © Urban Dictionary');
 
+		const language = await message.fetchLanguage();
```
Fetch language in run, pass it down

vladfrangu commented on src/commands/Tools/Websearch/country.ts:
```diff
@@ -20,27 +20,29 @@ const mapCurrency = (currency: CurrencyData) => `${currency.name} (${currency.sy
 	private async buildDisplay(message: KlasaMessage, countries: CountryResultOk) {
-		const titles = message.language.get(LanguageKeys.Commands.Tools.CountryTitles);
-		const fieldsData = message.language.get(LanguageKeys.Commands.Tools.CountryFields);
+		const language = await message.fetchLanguage();
```
Fetch lang in run, pass it down

vladfrangu commented on src/commands/Tools/Websearch/eshop.ts:
```diff
@@ -65,24 +65,25 @@ export default class extends RichDisplayCommand {
 
 	private async buildDisplay(entries: EShopHit[], message: KlasaMessage) {
-		const titles = message.language.get(LanguageKeys.Commands.Tools.EshopTitles);
+		const language = await message.fetchLanguage();
```
Ditto

vladfrangu commented on src/commands/Tools/Websearch/igdb.ts:
```diff
@@ -83,13 +83,14 @@ export default class extends RichDisplayCommand {
 	private async buildDisplay(entries: Game[], message: KlasaMessage) {
-		const titles = message.language.get(LanguageKeys.Commands.Tools.IgdbTitles);
-		const fieldsData = message.language.get(LanguageKeys.Commands.Tools.IgdbData);
+		const language = await message.fetchLanguage();
```
Ditto

vladfrangu commented on src/commands/Tools/Websearch/itunes.ts:
```diff
@@ -18,37 +18,38 @@ export default class extends RichDisplayCommand {
-		const { results: entries } = await this.fetchAPI(message, song);
-		if (!entries.length) throw message.language.get(LanguageKeys.System.NoResults);
+		const { results: entries } = await this.fetchAPI(language, song);
+		if (!entries.length) throw await message.fetchLocale(LanguageKeys.System.NoResults);
```
Already have language fetched, use it

vladfrangu commented on src/commands/Tools/Websearch/movies.ts:
```diff
@@ -23,14 +23,15 @@ export default class extends RichDisplayCommand {
 
 		const { results: entries } = await this.fetchAPI(message, movie, year);
-		if (!entries.length) throw message.language.get(LanguageKeys.System.NoResults);
+		if (!entries.length) throw await message.fetchLocale(LanguageKeys.System.NoResults);
```
Already have language fetched, use it

vladfrangu replied:
Chop chop

vladfrangu commented on src/commands/Tools/Websearch/movies.ts:
```diff
@@ -45,7 +46,7 @@ export default class extends RichDisplayCommand {
 			return await fetch<Tmdb.TmdbMovieList>(url, FetchResultTypes.JSON);
 		} catch {
-			throw message.language.get(LanguageKeys.System.QueryFail);
+			throw await message.fetchLocale(LanguageKeys.System.QueryFail);
```
pass language from run

vladfrangu commented on src/commands/Tools/Websearch/price.ts:
```diff
@@ -47,19 +48,19 @@ export default class extends SkyraCommand {
 			return body as CryptoCompareResultOk;
 		} catch {
-			throw message.language.get(LanguageKeys.Commands.Tools.PriceCurrencyNotFound);
+			throw await message.fetchLocale(LanguageKeys.Commands.Tools.PriceCurrencyNotFound);
```
pass language from run

vladfrangu commented on src/commands/Tools/Websearch/shows.ts:
```diff
@@ -23,14 +23,15 @@ export default class extends RichDisplayCommand {
 
 		const { results: entries } = await this.fetchAPI(message, show, year);
-		if (!entries.length) throw message.language.get(LanguageKeys.System.NoResults);
+		if (!entries.length) throw await message.fetchLocale(LanguageKeys.System.NoResults);
```
Use fetched language

vladfrangu commented on src/commands/Tools/Websearch/shows.ts:
```diff
@@ -45,7 +46,7 @@ export default class extends RichDisplayCommand {
 			return await fetch<Tmdb.TmdbSeriesList>(url, FetchResultTypes.JSON);
 		} catch {
-			throw message.language.get(LanguageKeys.System.QueryFail);
+			throw await message.fetchLocale(LanguageKeys.System.QueryFail);
```
Pass language down from run

vladfrangu commented on src/commands/Tools/Websearch/shows.ts:
```diff
@@ -56,13 +57,13 @@ export default class extends RichDisplayCommand {
 			return await fetch<Tmdb.TmdbSerie>(url, FetchResultTypes.JSON);
 		} catch {
-			throw message.language.get(LanguageKeys.System.QueryFail);
+			throw await message.fetchLocale(LanguageKeys.System.QueryFail);
```
Ditto

vladfrangu commented on src/commands/Twitch/followage.ts:
```diff
@@ -13,15 +13,17 @@ import { KlasaMessage } from 'klasa';
 	usageDelim: ' '
 })
 export default class extends SkyraCommand {
+	// TODO(QuantumlyTangled): Change KlasaMessage to GuildMessage
```
cmon.. just do it yourself

vladfrangu commented on src/commands/Twitch/twitch.ts:
```diff
@@ -13,15 +13,17 @@ import { KlasaMessage } from 'klasa';
 	usage: '<name:string>'
 })
 export default class extends SkyraCommand {
+	// TODO(QuantumlyTangled): Change KlasaMessage to GuildMessage
```
Ditto

vladfrangu commented on src/commands/Twitch/twitchsubscription.ts:
```diff
@@ -116,40 +121,35 @@ export default class extends SkyraCommand {
-			}
+				// Insert the entry to the database performing an upsert, if it created the entry, we tell the Twitch manager
+				// to send Twitch a message saying "hey, I want to be notified, can you pass me some data please?"
+				if (await this.upsertSubscription(message.guild!, streamer)) {
```
Migrate to GuildMessage

vladfrangu commented on src/commands/Twitch/twitchsubscription.ts:
```diff
@@ -203,15 +198,21 @@ export default class extends SkyraCommand {
+
 		// Retrieve all subscriptions for the guild
-		const subscriptions = message.guild!.settings.get(this.#kSettingsKey);
+		// const subscriptions = await message.guild.readSettings(this.#kSettingsKey);
```
So this isn't needed anymore?

vladfrangu commented on src/events/errors/commandError.ts:
```diff
@@ -1,47 +1,43 @@
-			} catch (err) {
-				return this.client.emit(Events.ApiError, err);
-			}
+			return message.alert(await message.fetchLocale(LanguageKeys.System.DiscordAborterror));
```
Fix casing

vladfrangu commented on src/events/guilds/members/guildMemberUpdateRolesNotify.ts:
```diff
@@ -0,0 +1,75 @@
+		const nextRoles = next.roles.cache;
+		if (prevRoles.equals(nextRoles)) return;
+
+		const [enabled, language] = await next.guild.readSettings((settings) => [
+			settings[GuildSettings.Events.MemberRoleUpdate],
+			settings.getLanguage()
+		]);
+
+		if (!enabled) return;
```
Check this BEFORE checking caches, as a member can have between 1 and like what, 250 roles? performance

vladfrangu commented on src/events/guilds/members/notMutedMemberAddJoinDirectMessage.ts:
```diff
@@ -0,0 +1,55 @@
+import { Guild, GuildMember, User } from 'discord.js';
+import { Event, EventOptions, Language } from 'klasa';
+
+const enum Matches {
```
Enum and regex are duplicated, please have only 1 in constants

vladfrangu commented on src/events/guilds/members/rawMemberRemoveFarewell.ts:
```diff
@@ -0,0 +1,47 @@
+	Member = '%MEMBER%',
+	MemberName = '%MEMBERNAME%',
+	MemberTag = '%MEMBERTAG%'
+}
```
MOAR DUPLICATES NYA

kyranet replied:
Actually, each one of those enums are different.

vladfrangu replied:
Yet they contain a bunch of similar if not identical tags...

vladfrangu commented on src/events/messages/guildMessageDeleteSnipe.ts:
```diff
@@ -0,0 +1,12 @@
+@ApplyOptions<EventOptions>({ event: Events.GuildMessageUpdate })
+export default class extends Event {
+	public run(message: GuildMessage) {
+		if (message.channel instanceof TextChannel) message.channel.sniped = message;
```
Use type checks

vladfrangu commented on src/events/messages/messageDelete.ts:
```diff
@@ -0,0 +1,9 @@
+export default class extends Event {
+	public run(message: KlasaMessage) {
+		if (message.partial || !message.guild || message.author.bot) return;
+		this.client.emit(Events.GuildMessageUpdate, message);
```
nit: deletes aren't message updates tyvm

vladfrangu commented on src/events/messages/rawMessageDelete.ts:
```diff
@@ -0,0 +1,19 @@
+
+export default class extends Event {
+	public constructor(store: EventStore, file: string[], directory: string) {
+		super(store, file, directory, { name: DiscordEvents.MessageDelete, emitter: store.client.ws });
```
You can replace your enum with the one in discord-api-types

vladfrangu commented on src/events/messages/rawMessageDeleteBulk.ts:
```diff
@@ -0,0 +1,19 @@
+
+export default class extends Event {
+	public constructor(store: EventStore, file: string[], directory: string) {
+		super(store, file, directory, { name: DiscordEvents.MessageDeleteBulk, emitter: store.client.ws });
```
Ditto

vladfrangu commented on src/events/messages/rawMessageDeleteBulkStarboard.ts:
```diff
@@ -0,0 +1,49 @@
+
+export default class extends Event {
+	public constructor(store: EventStore, file: string[], directory: string) {
+		super(store, file, directory, { name: DiscordEvents.MessageDeleteBulk, emitter: store.client.ws });
```
Ditto

vladfrangu commented on src/events/messages/rawMessageDeleteBulkStarboard.ts:
```diff
@@ -0,0 +1,49 @@
+
+		await api(this.client)
+			.channels(channel)
+			.messages['bulk-delete'].post({ data: { messages: filteredResults }, reason: 'Starboard Management: Message Deleted' })
```
This won't work for messages older than 14 days (will it be an issue idk)

kyranet replied:
Will never happen, because this event is fired on `MESSAGE_DELETE_BULK`, and starboard messages are always younger than that.

vladfrangu commented on src/events/messages/rawMessageDeleteStarboard.ts:
```diff
@@ -13,16 +12,7 @@ export default class extends Event {
 		super(store, file, directory, { name: DiscordEvents.MessageDelete, emitter: store.client.ws });
```
Ditto

vladfrangu commented on src/events/reactions/rawMessageReactionAdd.ts:
```diff
@@ -0,0 +1,43 @@
+
+export default class extends Event {
+	public constructor(store: EventStore, file: string[], directory: string) {
+		super(store, file, directory, { name: DiscordEvents.MessageReactionAdd, emitter: store.client.ws });
```
Ditto

vladfrangu commented on src/events/reactions/rawMessageReactionAdd.ts:
```diff
@@ -0,0 +1,43 @@
+				id: raw.emoji.id,
+				managed: Reflect.get(raw.emoji, 'managed') ?? null,
+				name: raw.emoji.name,
+				requireColons: 'require_colons' in raw.emoji ? raw.emoji.require_colons! : null,
```
Use Reflect.has for consistency

vladfrangu commented on src/events/reactions/rawReactionAddBlockList.ts:
```diff
@@ -0,0 +1,109 @@
+		if (!enabled || blockedReactions.length === 0 || ignoredChannels.includes(data.channel.id)) return;
+
+		const member = await data.guild.members.fetch(data.userID);
+		if (member.user.bot || this.hasPermissions(member)) return;
```
this.hasPermissions is a promise

vladfrangu commented on src/events/reactions/rawReactionAddBlockList.ts:
```diff
@@ -0,0 +1,109 @@
+			const points = typeof preProcessed === 'number' ? preProcessed : 1;
+			adder.add(data.userID, points);
+		} catch {
+			await this.processHardPunishment(data.guild, data.userID, await data.guild.readSettings(this.hardPunishmentPath.action));
```
read this in the main fetch of settings

vladfrangu commented on src/events/twitch/twitchStreamOffline.ts:
```diff
@@ -22,8 +21,12 @@ export default class extends Event {
+			const { allSubscriptions, language } = await guild.readSettings((settings) => ({
+				allSubscriptions: settings.notificationsStreamsTwitchStreamers,
+				language: settings.getLanguage()
+			}));
```
How come this doesn't use the `[SettingsKeys]` syntax?
```suggestion
			const [allSubscriptions, language] = await guild.readSettings((settings) => [
				settings.notificationsStreamsTwitchStreamers,
				settings.getLanguage()
			]);
```

vladfrangu commented on src/events/twitch/twitchStreamOnline.ts:
```diff
@@ -32,8 +31,12 @@ export default class extends Event {
+			const { allSubscriptions, language } = await guild.readSettings((settings) => ({
+				allSubscriptions: settings.notificationsStreamsTwitchStreamers,
+				language: settings.getLanguage()
+			}));
```
Ditto
```suggestion
			const [allSubscriptions, language] = await guild.readSettings((settings) => [
				settings.notificationsStreamsTwitchStreamers,
				settings.getLanguage()
			]);
```

vladfrangu commented on src/finalizers/commandAutoDelete.ts:
```diff
@@ -1,11 +1,10 @@
 		if (!message.guild) return;
 
-		const commandAutodelete = message.guild.settings.get(GuildSettings.CommandAutodelete);
+		const commandAutodelete = await message.guild.readSettings((guild) => guild.commandAutodelete);
```
nit: how come you don't use the `settings[]` syntax?

vladfrangu commented on src/lib/database/settings/Utils.ts:
```diff
@@ -0,0 +1,12 @@
+import { SchemaGroup } from './schema/SchemaGroup';
+import { SchemaKey } from './schema/SchemaKey';
+
+// TODO: Rename those to `isSchemaGroup` and `isSchemaKey`
```
So do it

vladfrangu commented on src/lib/database/settings/schema/SchemaKey.ts:
```diff
@@ -0,0 +1,130 @@
+	}
+
+	public get client(): SkyraClient {
+		return container.resolve<SkyraClient>('SkyraClient');
```
Any reason this doesn't pass the class directly?

kyranet replied:
This is built statically, before the client is instantiated.

vladfrangu commented on src/lib/database/settings/structures/collections/GuildSettingsCollection.ts:
```diff
@@ -0,0 +1,17 @@
+export class GuildSettingsCollection extends SettingsCollection<GuildEntity> {
+	public async fetch(key: string): Promise<GuildEntity> {
+		const connection = await DbSet.connect();
+		const existing = await connection.guilds.findOne(key);
```
```suggestion
		const { guilds } = await DbSet.connect();
		const existing = await guilds.findOne(key);
```

vladfrangu commented on src/lib/games/base/BaseReactionGame.ts:
```diff
@@ -27,7 +27,7 @@ export abstract class BaseReactionGame<T> extends BaseGame<T> {
 	protected async onStart(): Promise<unknown> {
 		try {
-			this.message = await this.message.send(pickRandom(this.message.language.get(LanguageKeys.System.Loading)), []);
+			this.message = await this.message.send(pickRandom(this.language.get(LanguageKeys.System.Loading)), []);
```
Random empty array?

vladfrangu commented on src/lib/misc/comparators.ts:
```diff
@@ -0,0 +1,5 @@
+export type Nullish = null | undefined;
+
+export function isNullish(value: unknown): value is Nullish {
```
Make this in @sapphire/utilities maybe?

kyranet replied:
On-going discussion: https://github.com/sapphire-project/utilities/issues/30

kyranet replied:
cc: @Favna

vladfrangu commented on src/lib/structures/ModerationCommand.ts:
```diff
@@ -124,48 +132,47 @@ export abstract class ModerationCommand<T = unknown> extends SkyraCommand {
-					? message.author
-					: null,
-			send: message.guild!.settings.get(GuildSettings.Messages.ModerationDM) && (await DbSet.fetchModerationDirectMessageEnabled(target.id))
+			moderator: 'no-author' in message.flagArgs ? null : Reflect.has(message.flagArgs, 'authored') || nameDisplay ? message.author : null,
```
Reflect.has in the first check

vladfrangu commented on src/lib/structures/WeebCommand.ts:
```diff
@@ -65,15 +64,16 @@ export abstract class WeebCommand extends SkyraCommand {
 			return await fetch<WeebCommandResult>(url, { headers: this.kHeaders }, FetchResultTypes.JSON);
 		} catch (unknownError: unknown) {
 			const error = unknownError as FetchError;
+			const language = await message.guild.fetchLanguage();
```
Pass language down from run

vladfrangu commented on src/lib/structures/managers/ModerationManager.ts:
```diff
@@ -47,16 +47,16 @@ export class ModerationManager extends Collection<number, ModerationEntity> {
-		return (channelID && (this.guild.channels.cache.get(channelID) as TextChannel)) || null;
+	public async fetchChannel() {
+		const channelID = await this.guild.readSettings(GuildSettings.Channels.ModerationLogs);
+		return channelID ? ((this.guild.channels.cache.get(channelID) ?? null) as TextChannel | null) : null;
```
This is an unreadable mess... 
```suggestion
		return this.guild.channels.cache.get(channelID) ?? null;
```

kyranet replied:
![image](https://user-images.githubusercontent.com/24852502/98568663-a5bb3f80-22b1-11eb-9a18-0ba7749cb9ea.png)

vladfrangu commented on src/lib/structures/managers/StarboardManager.ts:
```diff
@@ -46,16 +45,16 @@ export class StarboardManager extends Collection<string, StarboardEntity> {
-		return (channelID && (this.guild.channels.cache.get(channelID) as TextChannel)) || null;
+	public async getStarboardChannel() {
+		const channelID = await this.guild.readSettings(GuildSettings.Starboard.Channel);
+		return (channelID && (this.guild.channels.cache.get(channelID) as TextChannel)) ?? null;
```
```suggestion
		return this.guild.channels.cache.get(channelID) ?? null;
```

vladfrangu commented on src/lib/structures/managers/StarboardManager.ts:
```diff
@@ -46,16 +45,16 @@ export class StarboardManager extends Collection<string, StarboardEntity> {
 	 */
-	public get minimum() {
-		return this.guild.settings.get(GuildSettings.Starboard.Minimum);
+	public async getMinimumStars() {
```
useless async

vladfrangu commented on src/lib/util/API.ts:
```diff
@@ -1,26 +1,26 @@
 		typeof member !== 'undefined' &&
 		// If Roles.Admin is not configured, check MANAGE_GUILD, else check if the member has the role.
-		(roleID === null ? member.permissions.has(Permissions.FLAGS.MANAGE_GUILD) : member.roles.cache.has(roleID)) &&
+		(roleID === null || roleID === undefined ? member.permissions.has(Permissions.FLAGS.MANAGE_GUILD) : member.roles.cache.has(roleID)) &&
```
isNullish

vladfrangu commented on src/lib/util/Notifications/Twitch.ts:
```diff
@@ -1,3 +1,4 @@
+// What are you doing step-bro?
```
?

kyranet replied:
![image](https://user-images.githubusercontent.com/24852502/98569298-6e995e00-22b2-11eb-8153-d13e6c5febf0.png)

vladfrangu commented on src/lib/util/Security/ModerationActions.ts:
```diff
@@ -628,48 +631,42 @@ export class ModerationActions {
 			return true;
 		} catch (error) {
-			if (!(error instanceof DiscordAPIError)) throw this.guild.language.get(LanguageKeys.System.FetchbansFail);
+			if (!(error instanceof DiscordAPIError)) throw await this.guild.fetchLocale(LanguageKeys.System.FetchbansFail);
```
Casing fix

vladfrangu commented on src/lib/util/Security/ModerationActions.ts:
```diff
@@ -601,23 +608,19 @@ export class ModerationActions {
-			return Promise.reject(this.guild.language.get(LanguageKeys.Commands.Moderation.ActionSetupTooManyRoles));
+	public async restrictionSetup(message: KlasaMessage, path: ModerationSetupRestriction) {
+		const [roleID, language] = await this.guild.readSettings((settings) => [settings[path] as string | null, settings.getLanguage()]);
+		if (roleID !== null && this.guild.roles.cache.has(roleID)) throw language.get(LanguageKeys.Commands.Moderation.ActionSetupRestrictionExists);
```
use isNullish

vladfrangu commented on src/lib/util/Security/ModerationActions.ts:
```diff
@@ -679,25 +676,25 @@ export class ModerationActions {
 
 	private async fetchMe() {
-		return this.guild.me || this.guild.members.fetch(CLIENT_ID);
+		return this.guild.me ?? this.guild.members.fetch(CLIENT_ID);
```
nit: you could *just* return the `members.fetch(CLIENT_ID)`, since it'll return from cache if present

vladfrangu commented on src/lib/util/util.ts:
```diff
@@ -88,14 +88,14 @@ export async function streamToBuffer(stream: NodeJS.ReadableStream): Promise<Buf
-export function announcementCheck(message: Message) {
-	const announcementID = message.guild!.settings.get(GuildSettings.Roles.Subscriber);
-	if (!announcementID) throw message.language.get(LanguageKeys.Commands.Announcement.SubscribeNoRole);
+export async function announcementCheck(message: Message) {
```
Use GuildMessage

vladfrangu commented on src/monitors/Moderation/attachmentFilter.ts:
```diff
@@ -1,35 +1,45 @@
 		} catch (error) {
 			const points = (error as AdderError).amount;
+			const { maximum } = adder;
 			switch (action & 0b111) {
 				case 0b000:
```
Magic numbers 😭

kyranet replied:
Getting fixed with this system's rewrite in a following PR.

vladfrangu commented on src/monitors/Moderation/wordFilter.ts:
```diff
@@ -19,42 +19,38 @@ export default class extends ModerationMonitor {
 		if (content === null) return null;
 
-		return this.filter(removeConfusables(content), message.guild!.security.regexp!);
+		const regex = await message.guild.readSettings((settings) => settings.wordFilterRegExp);
```
Use `readSettings(SettingsKeys...)`

kyranet replied:
This actually isn't a settings key, it's computed.

vladfrangu commented on src/monitors/commandHandler.ts:
```diff
@@ -26,10 +25,13 @@ export default class extends Monitor {
 	public async sendPrefixReminder(message: KlasaMessage) {
 		if (message.guild !== null) {
-			const disabledChannels = message.guild.settings.get(GuildSettings.DisabledChannels);
+			const disabledChannels = await message.guild.readSettings((entity) => entity.disabledChannels);
```
Use `readSettings(SettingsKeys...)`

vladfrangu commented on src/monitors/imageLogs.ts:
```diff
@@ -17,6 +17,12 @@ const MAXIMUM_LENGTH = 1024 * 1024;
 
 export default class extends Monitor {
 	public async run(message: KlasaMessage) {
```
Use GuildMessage

vladfrangu commented on src/monitors/socialMonitor.ts:
```diff
@@ -15,6 +14,17 @@ export default class extends Monitor {
+			socialEnabled: settings[GuildSettings.Social.Enabled],
+			ignoredChannels: settings[GuildSettings.Social.IgnoreChannels],
+			multiplier: settings[GuildSettings.Social.Multiplier]
+		}));
```
```suggestion
		const [socialEnabled, ignoredChannels, multiplier] = await message.guild!.readSettings((settings) => [
			settings[GuildSettings.Social.Enabled],
			settings[GuildSettings.Social.IgnoreChannels],
			settings[GuildSettings.Social.Multiplier]
		]);
```

vladfrangu commented on src/monitors/socialMonitor.ts:
```diff
@@ -39,32 +48,36 @@ export default class extends Monitor {
+			achievementMessage: settings[GuildSettings.Social.AchieveMessage],
+			language: settings.getLanguage()
+		}));
+		if (shouldAchieve && message.channel.postable) {
```
```suggestion
		const [shouldAchieve, achievementMessage, language] = await message.guild!.readSettings((settings) => [
			settings[GuildSettings.Social.Achieve],
			settings[GuildSettings.Social.AchieveMessage],
			settings.getLanguage()
		]);

		if (shouldAchieve && message.channel.postable) {
```

vladfrangu commented on src/routes/guilds/guild/settings.ts:
```diff
@@ -43,15 +43,16 @@ export default class extends Route {
-			await botGuild.settings.update(entries, { arrayAction: 'overwrite', extraContext: { author: member.id } });
-			return response.json({ newSettings: botGuild.settings.toJSON() });
+			// await botGuild.settings.update(entries, { arrayAction: 'overwrite', extraContext: { author: member.id } });
+			return botGuild.readSettings((settings) => response.json(settings));
```
This will return a different object than before

vladfrangu commented on src/routes/oauth/logout.ts:
```diff
@@ -39,7 +39,6 @@ export default class extends Route {
 	}
 }
 
-// TODO(kyranet): remove cast once @vladfrangu adds OAuth data
```
If you remove this, might as well migrate to discord-api-types

vladfrangu commented on src/routes/oauth/oauthCallback.ts:
```diff
@@ -70,7 +70,6 @@ export default class extends Route {
 	}
 }
 
-// TODO(kyranet): remove cast once @vladfrangu adds OAuth data
```
Ditto, and stop duping types

vladfrangu commented on src/routes/oauth/oauthUser.ts:
```diff
@@ -170,7 +170,6 @@ export default class extends Route {
 	}
 }
 
-// TODO(kyranet): remove cast once @vladfrangu adds OAuth data
```
Ditto ditto ditto

vladfrangu commented on src/serializers/autorole.ts:
```diff
@@ -1,16 +1,32 @@
-		throw language.get(LanguageKeys.Serializers.AutoRoleInvalid);
+		const points = Number(rawPoints);
+		if (!Number.isSafeInteger(points)) {
+			// TODO(kyranet): Localize this.
```
Chop chop, SerializerError time

vladfrangu commented on src/serializers/boolean.ts:
```diff
@@ -1,22 +1,28 @@
-	private kFalses = new Set(['false', 'f', 'no', 'n', 'off', 'disable', 'disabled', '0', '-']);
+export default class UserSerializer extends Serializer<boolean> {
+	private readonly kTruthyTerms = new Set(['true', 't', 'yes', 'y', 'on', 'enable', 'enabled', '1', '+']);
+	private readonly kFalsyTerms = new Set(['false', 'f', 'no', 'n', 'off', 'disable', 'disabled', '0', '-']);
```
This is *fine*, but not i18n-able... For the future, consider changing that

vladfrangu commented on src/serializers/boolean.ts:
```diff
@@ -1,22 +1,28 @@
+	}
+
+	public stringify(data: boolean): string {
 		return data ? 'Enabled' : 'Disabled';
```
Dis too, but here you'd need more context

kyranet replied:
It has a context which comes with guild and language, it's fine. We can look this up later, added TODOs for this and the other.

vladfrangu commented on src/serializers/categoryortextchannel.ts:
```diff
@@ -1,28 +1,27 @@
+	public parse(value: string, context: SerializerUpdateContext): Awaited<string> {
+		const channel = context.entity.guild.channels.cache.get(value);
+		if (!channel) {
+			// TODO(kyranet): Localize this.
```
Chop~ Chop~

vladfrangu commented on src/serializers/categoryortextchannel.ts:
```diff
@@ -1,28 +1,27 @@
-		if (data.type === 'text' || data.type === 'category') return data;
-		throw language.get(LanguageKeys.Resolvers.InvalidChannel, { name: entry.key });
+	public stringify(data: string, context: SerializerUpdateContext): string {
+		return context.entity.guild.channels.cache.get(data)?.name ?? 'Unknown';
```
You..should probably add the ID too maybeidk
```suggestion
		return context.entity.guild.channels.cache.get(data)?.name ?? '#unknown-channel';
```

vladfrangu commented on src/serializers/channel.ts:
```diff
@@ -1,47 +1,52 @@
+	public parse(value: string, context: SerializerUpdateContext): Awaited<string> {
+		const channel = context.entity.guild.channels.cache.get(value);
+		if (!channel) {
+			// TODO(kyranet): Localize this.
```
Yes~

vladfrangu commented on src/serializers/commandautodelete.ts:
```diff
@@ -1,22 +1,36 @@
+	public parse(value: string): Awaited<CommandAutoDelete> {
+		const [command, rawDuration] = value.split(' ');
+		if (!command) {
+			throw new Error('Invalid command');
```
i18n

vladfrangu commented on src/serializers/customcommand.ts:
```diff
@@ -1,27 +1,50 @@
+
+export default class UserSerializer extends Serializer<CustomCommand> {
+	public parse(): Awaited<CustomCommand> {
+		throw new Error('Method not implemented.');
```
Intentional or?

vladfrangu commented on src/serializers/customcommand.ts:
```diff
@@ -1,27 +1,50 @@
+
+	public isValid(value: CustomCommand, context: SerializerUpdateContext): Awaited<boolean> {
+		if (typeof value.id !== 'string') {
+			throw new Error('The property "id" must be a string.');
```
Clearly you ahve some i18n'd keys, chop chop

vladfrangu commented on src/serializers/disabledcommandchannel.ts:
```diff
@@ -1,26 +1,57 @@
+	public isValid(value: DisabledCommandChannel, context: SerializerUpdateContext): Awaited<boolean> {
+		const channel = context.entity.guild.channels.cache.get(value.channel);
+		if (!channel) {
+			throw new Error('The channel does not exist.');
```
i18n

vladfrangu commented on src/serializers/disabledcommandchannel.ts:
```diff
@@ -1,26 +1,57 @@
+		const commands: string[] = [];
+		for (const command of commandIDs) {
+			if (!this.client.commands.has(command)) {
+				throw new Error(`The command \`${command}\` does not exist.`);
```
i18n

vladfrangu commented on src/serializers/disabledcommandchannel.ts:
```diff
@@ -1,26 +1,57 @@
+
+		for (const command of value.commands) {
+			if (!this.client.commands.has(command)) {
+				throw new Error(`The command \`${command}\` does not exist.`);
```
i18nnnn

vladfrangu commented on src/serializers/emoji.ts:
```diff
@@ -1,12 +1,25 @@
+	public parse(value: string, context: SerializerUpdateContext): Awaited<string> {
+		const resolved = resolveEmoji(value);
+		if (resolved === null) {
+			throw new Error(context.language.get(LanguageKeys.Resolvers.InvalidEmoji, { name: context.entry.name }));
```
why throw new error, and not throw directly

vladfrangu commented on src/serializers/emoji.ts:
```diff
@@ -1,12 +1,25 @@
+	public isValid(value: string, context: SerializerUpdateContext): Awaited<boolean> {
+		const resolved = resolveEmoji(value);
+		if (resolved === null) {
+			throw new Error(context.language.get(LanguageKeys.Resolvers.InvalidEmoji, { name: context.entry.name }));
```
ditto

vladfrangu commented on src/serializers/guild.ts:
```diff
@@ -0,0 +1,23 @@
+	public parse(value: string, context: SerializerUpdateContext): Awaited<string> {
+		const guild = this.client.guilds.cache.get(value);
+		if (guild) return guild.id;
+		throw context.language.get('resolverInvalidGuild', { name: context.entry.name });
```
Keys not using LanguageKeys namespace..

vladfrangu commented on src/serializers/guild.ts:
```diff
@@ -0,0 +1,23 @@
+	public isValid(value: string, context: SerializerUpdateContext): Awaited<boolean> {
+		const guild = this.client.guilds.cache.get(value);
+		if (!guild) {
+			throw context.language.get('resolverInvalidGuild', { name: context.entry.name });
```
bonk

vladfrangu commented on src/serializers/number.ts:
```diff
@@ -0,0 +1,43 @@
+			case 'integer': {
+				const number = parseInt(value, 10);
+				if (Number.isInteger(number) && Serializer.minOrMax(number, context)) return number;
+				throw context.language.get('resolverInvalidInt', { name: context.entry.name });
```
Keys not used with LanguageKeys

vladfrangu commented on src/serializers/number.ts:
```diff
@@ -0,0 +1,43 @@
+			case 'float': {
+				const number = parseFloat(value);
+				if (!Number.isNaN(number) && Serializer.minOrMax(number, context)) return number;
+				throw context.language.get('resolverInvalidFloat', { name: context.entry.name });
```
^

vladfrangu commented on src/serializers/number.ts:
```diff
@@ -0,0 +1,43 @@
+		switch (context.entry.type) {
+			case 'integer': {
+				if (typeof value === 'number' && Number.isInteger(value) && Serializer.minOrMax(value, context)) return true;
+				throw context.language.get('resolverInvalidInt', { name: context.entry.name });
```
*ahem*

vladfrangu commented on src/serializers/permissionnode.ts:
```diff
@@ -1,34 +1,36 @@
-		if (guild === null) throw new TypeError('guild must not be null.');
+export default class UserSerializer extends Serializer<PermissionsNode> {
+	public parse(): Awaited<PermissionsNode> {
+		throw new Error('Method not implemented.');
```
Intentional?

vladfrangu commented on src/serializers/piece.ts:
```diff
@@ -0,0 +1,31 @@
+
+	public parse(value: string, { entry, language }: SerializerUpdateContext): Awaited<string> {
+		const store = this.client.pieceStores.get(`${entry.type}s`);
+		if (!store) throw language.get('resolverInvalidStore', { store: entry.type });
```
Not using LanguageKeys enum, won't repeat myself for the other 3 instances

vladfrangu commented on src/serializers/piece.ts:
```diff
@@ -0,0 +1,31 @@
+import { Awaited } from '@sapphire/utilities';
+
+export default class UserSerializer extends Serializer<string> {
+	public constructor(store: SerializerStore, file: string[], directory: string) {
+		super(store, file, directory);
+
+		// Adds all pieces, custom or not, to this serialize piece for use in schemas
+		this.aliases = [...this.client.pieceStores.keys()].map((type) => type.slice(0, -1));
+	}
```
class field for aliases

kyranet replied:
... what?

vladfrangu commented on src/serializers/reactionrole.ts:
```diff
@@ -1,28 +1,31 @@
-	public validate(data: ReactionRole, { language }: SerializerUpdateContext) {
+export default class UserSerializer extends Serializer<ReactionRole> {
+	public parse(): Awaited<ReactionRole> {
+		// TODO (kyranet): implement this
```
Chopchop

vladfrangu commented on src/serializers/role.ts:
```diff
@@ -0,0 +1,20 @@
+		const id = UserSerializer.regex.role.exec(value);
+		const role = id ? context.entity.guild.roles.cache.get(id[0]) : context.entity.guild.roles.cache.find((r) => r.name === value);
+		if (role) return role.id;
+		throw context.language.get('resolverInvalidRole', { name: context.entry.name });
```
Not using LanguageKeys namespace

vladfrangu commented on src/serializers/stickyrole.ts:
```diff
@@ -1,26 +1,28 @@
-	public validate(data: StickyRole, { language, guild }: SerializerUpdateContext) {
+export default class UserSerializer extends Serializer<StickyRole> {
+	public parse(): Awaited<StickyRole> {
+		throw new Error('Method not implemented.');
```
Intentional?

vladfrangu commented on src/serializers/string.ts:
```diff
@@ -0,0 +1,13 @@
+
+export default class UserSerializer extends Serializer<string> {
+	public parse(value: string, context: SerializerUpdateContext): Awaited<string> {
+		Serializer.minOrMax(value.length, context);
```
?

vladfrangu commented on src/serializers/triggeralias.ts:
```diff
@@ -1,13 +1,16 @@
-		if (isObject(data) && Object.keys(data).length === 2 && typeof data.input === 'string' && typeof data.output === 'string') return data;
+export default class UserSerializer extends Serializer<TriggerAlias> {
+	public parse(): Awaited<TriggerAlias> {
+		throw new Error('Method not implemented.');
```
Intentional?

vladfrangu commented on src/serializers/triggerinclude.ts:
```diff
@@ -1,18 +1,22 @@
-	public validate(data: TriggerIncludes, { language }: SerializerUpdateContext) {
+export default class UserSerializer extends Serializer<TriggerIncludes> {
+	public parse(): Awaited<TriggerIncludes> {
+		// TODO (kyranet): implement this
```
Dew it

vladfrangu commented on src/serializers/twitchsubscription.ts:
```diff
@@ -1,9 +1,13 @@
-	public validate(data: NotificationsStreamTwitch, { language }: SerializerUpdateContext) {
+export default class UserSerializer extends Serializer<NotificationsStreamTwitch> {
+	public parse(): Awaited<NotificationsStreamTwitch> {
+		throw new Error('Method not implemented.');
```
Intentional?

vladfrangu commented on src/serializers/uniqueroleset.ts:
```diff
@@ -1,26 +1,29 @@
-	public validate(data: UniqueRoleSet, { language, guild }: SerializerUpdateContext) {
+export default class UserSerializer extends Serializer<UniqueRoleSet> {
+	public parse(): Awaited<UniqueRoleSet> {
+		// TODO (kyranet): implement this
```
You said you wanted things to do, watcha waiting for?

vladfrangu commented on src/serializers/url.ts:
```diff
@@ -1,17 +1,37 @@
+		try {
+			const { hostname } = new URL(this.kProtocol.test(value) ? value : `https://${value}`);
+			return hostname.length <= 128;
+		} catch (error) {
```
```suggestion
		} catch {
```

vladfrangu commented on src/serializers/url.ts:
```diff
@@ -1,17 +1,37 @@
+			if (hostname.length <= 128) return hostname;
+			throw new Error('💥');
+		} catch {
+			throw new Error(language.get(LanguageKeys.Resolvers.MinmaxMaxInclusive, { name: entry.name, max: 128 }));
```
Any reason to throw a new Error?

vladfrangu commented on src/serializers/url.ts:
```diff
@@ -1,17 +1,37 @@
-			return Promise.reject(language.get(LanguageKeys.Resolvers.InvalidUrl, { name: entry.path }));
+			return Promise.reject(language.get(LanguageKeys.Resolvers.InvalidUrl, { name: entry.name }));
 		}
 	}
```
Modernise: async/await

vladfrangu commented on src/serializers/user.ts:
```diff
@@ -0,0 +1,27 @@
+		const id = Serializer.regex.userOrMember.exec(value);
+		const user = id ? await this.client.users.fetch(id[1]).catch(() => null) : null;
+		if (user) return user.id;
+		throw context.language.get('resolverInvalidUser', { name: context.entry.name });
```
Key not using LanguageKeys namespace
Same lower

vladfrangu commented on src/serializers/user.ts:
```diff
@@ -0,0 +1,27 @@
+	}
+
+	public stringify(value: string) {
+		return this.client.users.cache.get(value)?.username ?? value;
```
*Probably* should return tag instead

vladfrangu reviewed (changes requested):
Overall, this PR is.. *fine*. There are a LOT of inconsistencies between how you fetch languages, sometimes with fetchLocale, sometimes with sendLocale, sometimes fetching the whole object, sometimes even mixing 2 or more kinds. This sort inconsistency is not ideal.

Try looking into hydrating the language object on the message itself (since 99% of the time you'll use it anyways), so you still have message.language. Otherwise, choose how you'll do languages, and stick by it as much as possible!

kyranet committed:
fix: requested changes, part I

kyranet committed:
fix: resolved build errors

kyranet committed:
fix: requested changes, part II

vladfrangu commented on src/commands/GameIntegration/overwatch.ts:
```diff
@@ -178,8 +183,8 @@ export default class extends RichDisplayCommand {
 	/** Extracts top heroes from overwatchData for either competitive play or quickplay and returns it in a format valid for `MessageEmbed` description */
-	private extractTopHeroes(
-		message: KlasaMessage,
+	private async extractTopHeroes(
```
Any reason for async? If yes, then there's code to be fixed at line 102-103

kyranet committed:
fix: requested changes, part III

favna committed:
chore: fixes

vladfrangu commented on src/commands/Management/Configuration/manageCommandAutodelete.ts:
```diff
@@ -26,62 +26,69 @@ import { KlasaMessage } from 'klasa';
-		const commandAutodelete = message.guild!.settings.get(GuildSettings.CommandAutodelete);
-		if (!commandAutodelete.length) throw message.language.get(LanguageKeys.Commands.Management.ManageCommandAutoDeleteShowEmpty);
+	public async show(message: GuildMessage) {
+		const commandAutodelete = await message.guild.readSettings(GuildSettings.CommandAutodelete);
```
Casing

vladfrangu commented on src/commands/Management/Configuration/manageCommandAutodelete.ts:
```diff
@@ -26,62 +26,69 @@ import { KlasaMessage } from 'klasa';
+	public async reset(message: GuildMessage) {
+		const language = await message.guild.writeSettings((settings) => {
+			const values = settings[GuildSettings.CommandAutodelete];
+			values.length = 0;
```
```suggestion
			settings[GuildSettings.CommandAutodelete].length = 0
```

vladfrangu commented on src/commands/Management/Configuration/setIgnoreChannels.ts:
```diff
@@ -21,19 +21,33 @@ export default class extends SkyraCommand {
-			extraContext: { author: message.author.id }
+		else if (!isTextBasedChannel(channel)) throw await message.fetchLocale(LanguageKeys.Misc.ConfigurationTextChannelRequired);
+
+		const [language, oldLength, newLength] = await message.guild!.writeSettings((settings) => {
```
Use GuildMessage

kyranet committed:
fix: requested changes, part IV

kyranet committed:
fix: resolved path circulars

favna committed:
chore: more fixes

kyranet committed:
fix: add fetchPrefix and fetchLanguage in SkyraClient

kyranet committed:
fix: resolved GSC not setting the entry into the cache

favna committed:
chore: language keys

kyranet committed:
fix: more bugfixes

favna committed:
chore: serializer i18n

kyranet committed:
fix: resolved a pair of critical bugs

IT'S ALIVE!!

favna committed:
chore: code cleanup

favna committed:
chore: cleanup unncessary type assertions

favna committed:
chore: formatting all files

favna committed:
chore: resolve some TODOs

favna committed:
chore: bump packages

kyranet committed:
fix: I definitely did not forget to add this

favna committed:
feat: add field move effects for

favna committed:
chore: thanks DefinitelyTyped

favna committed:
chore: move table keys to a better folder

favna committed:
chore: fix migrations and add commandUses nuke migration

favna committed:
fix: fixes

Co-authored-by: Antonio Román <kyradiscord@gmail.com>

favna committed:
chore: kyra's shit and klasa upgrades

kyranet committed:
fix: serializers not loading

kyranet committed:
fix: per-settings lock, fixed Schema{Group,Key}

kyranet committed:
chore: updated Klasa for the wiggles

kyranet committed:
feat: completed SettingsMenu and PATCH settings route

Also changes POST to PATCH for correctness

kyranet committed:
fix: translate raw keys to entity properties

vladfrangu commented on src/arguments/song.ts:
```diff
@@ -177,12 +178,13 @@ export default class extends Argument {
+			GuildSettings.Music.MaximumDuration,
+			GuildSettings.Music.AllowStreams
+		]);
 		const filteredStreams = allowStreams ? filter(tracks.values(), (track) => !track.info.isStream) : tracks.values();
```
isn't this inverted? If we allow streams, allow just non-streams, else return all?

vladfrangu commented on src/commands/Admin/roleset.ts:
```diff
@@ -36,110 +36,90 @@ import { KlasaMessage } from 'klasa';
-				}
-			]);
-		}
+		const { created, roleSets } = await message.guild.writeSettings((settings) => {
```
Well I can't write a suggestion but fix this `{}` => `[]`

vladfrangu commented on src/commands/Admin/roleset.ts:
```diff
@@ -36,110 +36,90 @@ import { KlasaMessage } from 'klasa';
-		});
-		return message.sendLocale(LanguageKeys.Commands.Admin.RolesetAdded, [
-			{
+		const language = await message.fetchLanguage();
```
return from writeSettings

vladfrangu commented on src/commands/Admin/roleset.ts:
```diff
@@ -36,110 +36,90 @@ import { KlasaMessage } from 'klasa';
-		return message.sendLocale(LanguageKeys.Commands.Admin.RolesetRemoved, [
-			{
+
+		const language = await message.fetchLanguage();
```
Return from writeSettings

vladfrangu commented on src/commands/Developers/yarn.ts:
```diff
@@ -34,42 +33,44 @@ export default class extends SkyraCommand {
 	public async run(message: KlasaMessage, [pkg]: [string]) {
+		const language = await message.fetchLanguage();
 		const response = await message.sendEmbed(
-			new MessageEmbed().setDescription(pickRandom(message.language.get(LanguageKeys.System.Loading))).setColor(BrandingColors.Secondary)
+			new MessageEmbed().setDescription(pickRandom(language.get(LanguageKeys.System.Loading))).setColor(BrandingColors.Secondary)
 		);
```
TODO separate PR: make this a util function; add TODO comment

vladfrangu commented on src/commands/Developers/yarn.ts:
```diff
@@ -34,42 +33,44 @@ export default class extends SkyraCommand {
 	}
 
 	private async fetchApi(message: KlasaMessage, pkg: string) {
+		const language = await message.fetchLanguage();
```
Pass down from run

favna committed:
chore: dependencies go brrrr

favna committed:
chore: "fix" rate

kyranet committed:
fix: resolved all serializer errors

kyranet committed:
chore: downgrade TypeORM

0.2.29 has broken relationships on PGSQL, **DO NOT USE**.

I repeat, **DO NOT USE**!

kyranet committed:
fix: resolved most TODOs

kyranet committed:
feat: localize bool resolver

kyranet committed:
fix: resolved all piece errors

vladfrangu commented on src/commands/Fun/rate.ts:
```diff
@@ -14,16 +15,23 @@ import { KlasaMessage } from 'klasa';
-			[ratewaifu, user] = message.language.get(LanguageKeys.Commands.Fun.RateMyself);
+			[ratewaifu, user] = language.get(LanguageKeys.Commands.Fun.RateMyself);
+		} else if (this.devRegex.test(user)) {
+			rate = 101;
```
Hoe where's me 🤔

kyranet replied:
That was written by @Favna.

kyranet committed:
fix: removed useless TODO

vladfrangu commented on src/commands/Games/higherLower.ts:
```diff
@@ -70,6 +70,7 @@ export default class extends SkyraCommand {
 				}
 			),
 			response,
+			language: await message.fetchLanguage(),
```
Fetch this above, re-use in throw

vladfrangu commented on src/commands/Management/Configuration/manageCommandAutodelete.ts:
```diff
@@ -26,62 +26,68 @@ import { KlasaMessage } from 'klasa';
-		if (!commandAutodelete.length) throw message.language.get(LanguageKeys.Commands.Management.ManageCommandAutoDeleteShowEmpty);
+	public async show(message: GuildMessage) {
+		const commandAutoDelete = await message.guild.readSettings(GuildSettings.CommandAutodelete);
+		const language = await message.fetchLanguage();
```
Unify like in other files

vladfrangu commented on src/commands/Management/Configuration/manageCommandAutodelete.ts:
```diff
@@ -26,62 +26,68 @@ import { KlasaMessage } from 'klasa';
-		const value: readonly [string, number] = [channel.id, duration];
+	public async add(message: GuildMessage, [channel, duration]: [TextChannel, number]) {
+		const language = await message.guild.writeSettings((settings) => {
+			const commandAutodelete = settings[GuildSettings.CommandAutodelete];
```
CASING

vladfrangu commented on src/commands/Management/Configuration/manageReactionRoles.ts:
```diff
@@ -48,94 +47,110 @@ import { ArrayActions, KlasaMessage } from 'klasa';
+			return message.send(
+				language.get(LanguageKeys.Commands.Management.ManageReactionRolesAddChannel, {
+					emoji: displayEmoji(reactionRole.emoji),
+					channel: `<#${channel}>`
```
Since channel is a... TextChannel, this will create `<#<#id>>` right?

vladfrangu commented on src/commands/Management/createMute.ts:
```diff
@@ -1,49 +1,48 @@
+	description: (language) => language.get(LanguageKeys.Commands.Management.CreateMuteDescription),
+	extendedHelp: (language) => language.get(LanguageKeys.Commands.Management.CreateMuteExtended),
+	permissionLevel: PermissionLevels.Administrator,
+	requiredGuildPermissions: ['MANAGE_CHANNELS', 'MANAGE_ROLES'],
```
Use permissions flags from d.js?

vladfrangu commented on src/commands/Management/roles.ts:
```diff
@@ -140,25 +141,26 @@ export default class extends RichDisplayCommand {
-		if (!roles.length) throw message.language.get(LanguageKeys.Commands.Management.RolesListEmpty);
+		if (!roles.length) throw await message.fetchLocale(LanguageKeys.Commands.Management.RolesListEmpty);
 
+		const language = await message.fetchLanguage();
```
fetch before throw :3

vladfrangu commented on src/commands/Management/triggers.ts:
```diff
@@ -22,80 +21,87 @@ const REG_TYPE = /^(alias|reaction)$/i;
-	}
+			} else if (type === 'alias') {
+				const command = message.client.commands.get(arg);
+				if (command && command.permissionLevel < 10) return command.name;
```
Use PermissionLevel enum

vladfrangu commented on src/commands/Management/triggers.ts:
```diff
@@ -22,80 +21,87 @@ const REG_TYPE = /^(alias|reaction)$/i;
+				throw await message.fetchLocale(LanguageKeys.Commands.Management.TriggersInvalidAlias);
+			} else {
+				return null;
+			}
```
this can drop the else, keep just the return

vladfrangu commented on src/commands/Management/triggers.ts:
```diff
@@ -22,80 +21,87 @@ const REG_TYPE = /^(alias|reaction)$/i;
+				if (!emoji) throw await message.fetchLocale(LanguageKeys.Commands.Management.TriggersInvalidReaction);
+
+				try {
+					await message.react(emoji);
```
This..is not a valid test per say. With perms denied, this WILL throw even if the emoji is valid. What you want is to check if the emoji exists at all (probs HEADing the CDN)

kyranet replied:
It's more whether or not Skyra can use the emoji.

vladfrangu replied:
~~Preeetty sure bots can now use any emojis~~ Only webhooks

vladfrangu commented on src/commands/Moderation/Utilities/time.ts:
```diff
@@ -1,37 +1,39 @@
-		this.createCustomResolver('timer', async (arg, possible, message, [cancel]) => {
+@ApplyOptions<SkyraCommandOptions>({
+	cooldown: 5,
+	description: 'Sets a timer.',
```
i18n.

vladfrangu commented on src/commands/Moderation/unwarn.ts:
```diff
@@ -18,20 +17,21 @@ export default class extends ModerationCommand {
-		if (message.guild!.settings.get(GuildSettings.Messages.ModerationMessageDisplay)) {
-			const originalReason = message.guild!.settings.get(GuildSettings.Messages.ModerationReasonDisplay) ? unwarnLog.reason : null;
+		if (await message.guild.readSettings(GuildSettings.Messages.ModerationMessageDisplay)) {
+			const originalReason = (await message.guild.readSettings(GuildSettings.Messages.ModerationReasonDisplay)) ? unwarnLog.reason : null;
```
Unify all fetches in one call

vladfrangu commented on src/commands/Moderation/unwarn.ts:
```diff
@@ -42,14 +42,14 @@ export default class extends ModerationCommand {
 					}
 				]
-			);
+			) as Promise<GuildMessage>;
```
Is this cast needed?

kyranet replied:
Yes.

vladfrangu commented on src/commands/Music/queue.ts:
```diff
@@ -49,24 +51,21 @@ export default class extends MusicCommand {
 			if (!track.isStream) {
 				nowPlayingDescription.push(
-					message.language.get(LanguageKeys.Commands.Music.QueueNowplayingTimeRemaining, {
+					language.get(LanguageKeys.Commands.Music.QueueNowplayingTimeRemaining, {
```
Casing

vladfrangu commented on src/commands/Music/queue.ts:
```diff
@@ -39,8 +41,8 @@ export default class extends MusicCommand {
 			const nowPlayingDescription = [
-				track.isStream ? message.language.get(LanguageKeys.Commands.Music.QueueNowplayingLiveStream) : showSeconds(track.length),
-				message.language.get(LanguageKeys.Commands.Music.QueueNowplaying, {
+				track.isStream ? language.get(LanguageKeys.Commands.Music.QueueNowplayingLiveStream) : showSeconds(track.length),
```
Casing

vladfrangu commented on src/commands/Music/skip.ts:
```diff
@@ -25,23 +25,23 @@ export default class extends MusicCommand {
 		const track = await audio.getCurrentTrack();
 		await audio.next({ skipped: true });
-		this.client.emit(Events.MusicSongSkipNotify, message, track!);
+		return this.client.emit(Events.MusicSongSkipNotify, message, track);
```
return?

vladfrangu commented on src/commands/Pokemon/ability.ts:
```diff
@@ -19,32 +19,35 @@ import { KlasaMessage } from 'klasa';
+			);
+
+		if (abilityDetails.isFieldAbility) {
+			embed.fields.unshift({ name: embedTitles.fieldEffectTitle, value: abilityDetails.isFieldAbility, inline: false });
```
nit: Could also use spliceFields

vladfrangu commented on src/commands/Social/Profile Management/banner.ts:
```diff
@@ -147,16 +149,17 @@ export default class extends SkyraCommand {
 		const user = await users.ensureProfile(message.author.id);
 		const banners = new Set(user.profile.banners);
-		if (!banners.size) throw message.language.get(LanguageKeys.Commands.Social.BannerUserlistEmpty, { prefix });
+		if (!banners.size) throw language.get(LanguageKeys.Commands.Social.BannerUserlistEmpty, { prefix });
```
UserList

vladfrangu commented on src/commands/Social/autorole.ts:
```diff
@@ -1,102 +1,128 @@
+export default class extends SkyraCommand {
+	public async init() {
+		this.customizeResponse('role', async (message) => message.fetchLocale(LanguageKeys.Misc.CommandRequireRole)) //
+			.customizeResponse('points', async (message) => message.fetchLocale(LanguageKeys.Commands.Social.AutoRolePointsRequired)); //
```
useless async keyword

vladfrangu commented on src/commands/Social/daily.ts:
```diff
@@ -69,7 +68,7 @@ export default class extends SkyraCommand {
 			await this.client.schedules.add(Schedules.Reminder, nextTime, {
 				data: {
-					content: message.language.get(LanguageKeys.Commands.Social.DailyCollect),
+					content: await message.fetchLocale(LanguageKeys.Commands.Social.DailyCollect),
```
pass language down

vladfrangu commented on src/commands/Social/pay.ts:
```diff
@@ -61,16 +62,16 @@ export default class extends SkyraCommand {
 
-	private denyPayment(message: KlasaMessage) {
-		return message.sendMessage(message.language.get(LanguageKeys.Commands.Social.PayPromptDeny));
+	private async denyPayment(message: GuildMessage, language: Language) {
```
useless async

vladfrangu commented on src/commands/Starboard/star.ts:
```diff
@@ -31,13 +29,13 @@ const MEDALS = ['🥇', '🥈', '🥉'];
-	public async random(message: KlasaMessage, [member]: [GuildMember?]): Promise<KlasaMessage | KlasaMessage[]> {
-		const minimum = message.guild!.settings.get(GuildSettings.Starboard.Minimum);
+	public async random(message: GuildMessage, [member]: [GuildMember?]): Promise<Message | Message[]> {
+		const minimum = await message.guild.readSettings(GuildSettings.Starboard.Minimum);
```
Return language, adapt negatives with throw language.get for consistency with other files

vladfrangu commented on src/commands/Starboard/star.ts:
```diff
@@ -80,21 +76,21 @@ export default class extends SkyraCommand {
-	public async top(message: KlasaMessage, [member, timespan]: [GuildMember?, number?]) {
-		const minimum = message.guild!.settings.get(GuildSettings.Starboard.Minimum);
+	public async top(message: GuildMessage, [member, timespan]: [GuildMember?, number?]) {
+		const minimum = await message.guild.readSettings(GuildSettings.Starboard.Minimum);
```
Ditto as above

vladfrangu commented on src/commands/Suggestions/resolvesuggestion.ts:
```diff
@@ -33,24 +33,26 @@ const enum SuggestionsColors {
 			// Validate the suggestions channel ID
-			const channelID = message.guild!.settings.get(GuildSettings.Suggestions.SuggestionsChannel);
-			if (!channelID) throw message.language.get(LanguageKeys.Commands.Suggestions.SuggestNoSetup, { username: message.author.username });
+			const channelID = await message.guild!.readSettings(GuildSettings.Suggestions.SuggestionsChannel);
```
Unite

vladfrangu commented on src/commands/Suggestions/resolvesuggestion.ts:
```diff
@@ -64,29 +66,31 @@ const enum SuggestionsColors {
-		const author = await this.getAuthor(message, shouldHideAuthor);
-		const actions = message.language.get(LanguageKeys.Commands.Suggestions.ResolveSuggestionActions, { author });
-		const DMActions = message.language.get(LanguageKeys.Commands.Suggestions.ResolveSuggestionActionsDms, { author, guild: message.guild!.name });
+		const language = await message.fetchLanguage();
```
Unite with readSettings

vladfrangu commented on src/commands/Tags/tag.ts:
```diff
@@ -26,90 +25,95 @@ import { CommandOptions, KlasaMessage } from 'klasa';
+			settings[GuildSettings.CustomCommands].push(this.createTag(message, id, content));
 		});
 
 		return message.sendLocale(LanguageKeys.Commands.Tags.TagAdded, [{ name: id, content: cutText(content, 1850) }]);
```
Bruh, re-use language

vladfrangu commented on src/commands/Tags/tag.ts:
```diff
@@ -26,90 +25,95 @@ import { CommandOptions, KlasaMessage } from 'klasa';
+			settings[GuildSettings.CustomCommands].splice(tagIndex, 1);
 		});
 
 		return message.sendLocale(LanguageKeys.Commands.Tags.TagRemoved, [{ name: id }]);
```
DItto

vladfrangu commented on src/commands/Tags/tag.ts:
```diff
@@ -26,90 +25,95 @@ import { CommandOptions, KlasaMessage } from 'klasa';
+			settings[GuildSettings.CustomCommands].splice(tagIndex, 1, this.createTag(message, id, content));
 		});
 
 		return message.sendLocale(LanguageKeys.Commands.Tags.TagEdited, [{ name: id, content: cutText(content, 1000) }]);
```
Ditto

vladfrangu commented on src/commands/Tools/topinvites.ts:
```diff
@@ -19,32 +20,33 @@ import { KlasaMessage, Timestamp } from 'klasa';
+	public async run(message: GuildMessage) {
 		const response = await message.sendEmbed(
-			new MessageEmbed().setDescription(pickRandom(message.language.get(LanguageKeys.System.Loading))).setColor(BrandingColors.Secondary)
+			new MessageEmbed().setDescription(pickRandom(await message.fetchLocale(LanguageKeys.System.Loading))).setColor(BrandingColors.Secondary)
```
Fetch language onceeee

vladfrangu commented on src/commands/Tools/topinvites.ts:
```diff
@@ -19,32 +20,33 @@ import { KlasaMessage, Timestamp } from 'klasa';
 
-	private async buildDisplay(message: KlasaMessage, invites: NonNullableInvite[]) {
+	private async buildDisplay(message: GuildMessage, invites: NonNullableInvite[]) {
+		const language = await message.fetchLanguage();
```
Pass down from run

vladfrangu commented on src/commands/Tools/whois.ts:
```diff
@@ -1,13 +1,27 @@
+	runIn: ['text'],
+	usage: '(user:username)'
+})
+@CreateResolvers([
+	['username', (arg, possible, message) => (arg ? message.client.arguments.get('username')!.run(arg, possible, message) : message.author)]
+])
```
Ok but is there any reason for this? Just make it optional and default `user = message.author` in the run block

kyranet replied:
The behaviour is completely different, `s!whois #somechannel` for example should give an error, not default to the author.

vladfrangu commented on src/commands/Twitch/followage.ts:
```diff
@@ -42,14 +42,14 @@ export default class extends SkyraCommand {
+			if (!data || data.length < 2) throw language.get(LanguageKeys.Commands.Twitch.FollowageMissingEntries);
 
 			return data;
 		} catch (err) {
```
```suggestion
		} catch {
```

vladfrangu commented on src/commands/Twitch/twitchsubscription.ts:
```diff
@@ -159,41 +159,36 @@ export default class extends SkyraCommand {
-		const subscriptions = message.guild!.settings.get(this.#kSettingsKey);
-		const subscriptionIndex = subscriptions.findIndex((sub) => sub[0] === streamer.id);
+	public async remove(message: GuildMessage, [streamer, channel, status]: [Streamer, Channel, Status]) {
+		const language = await message.fetchLanguage();
```
fetch in writeSettings, and re-use from there

vladfrangu commented on src/events/errors/commandError.ts:
```diff
@@ -1,47 +1,43 @@
 	// Unknown Channel
 	10003,
 	// Unknown Message
 	10008
```
nit: Migrate to discord-api-types resterrorcodes?

vladfrangu commented on src/events/errors/commandError.ts:
```diff
@@ -1,47 +1,43 @@
-	public async run(message: KlasaMessage, command: Command, _: string[], error: string | Error) {
+	public async run(message: KlasaMessage, command: Command, _: string[], error: Awaited<string> | Error) {
+		// Verify if the error is promise or not, if it is then await that promise
+		if (isThenable(error)) error = await error;
```
*ironic, this doesn't seem to work*

kyranet replied:
What do you mean?

vladfrangu replied:
Fav showed in the dev server, throwing without await returned `[object promise]`

vladfrangu commented on src/events/guilds/members/rawMemberRemoveNotify.ts:
```diff
@@ -0,0 +1,79 @@
+
+@ApplyOptions<EventOptions>({ name: Events.RawMemberRemove })
+export default class extends Event {
+	public async run(guild: Guild, data: GatewayGuildMemberRemoveDispatch['d']) {
```
considering all you use is the user obj, destructure it

vladfrangu commented on src/events/guilds/members/rawMemberRemoveNotify.ts:
```diff
@@ -0,0 +1,79 @@
+		this.client.emit(Events.GuildMessageLog, MessageLogsEnum.Member, guild, () =>
+			new MessageEmbed()
+				.setColor(Colors.Red)
+				.setAuthor(`${data.user.username}#${data.user.discriminator} (${data.user.id})`, getDisplayAvatar(data.user.id, data.user))
```
Maybe you should look into `client.users.add(data.user, false)` and then using user.displayAvatarURL / user.tag?

kyranet replied:
No, that'd needlessly duplicate ops.

vladfrangu commented on src/events/guilds/rawGuildDelete.ts:
```diff
@@ -0,0 +1,13 @@
+	}
+
+	public run(data: GatewayGuildDeleteDispatch['d']) {
+		this.client.settings.guilds.delete(data.id);
+		this.client.audio.queues?.delete(data.id);
```
You aren't checking if the guild just went unavailable...

vladfrangu commented on src/events/messages/rawMessageDeleteBulkStarboard.ts:
```diff
@@ -0,0 +1,48 @@
+
+export default class extends Event {
+	public constructor(store: EventStore, file: string[], directory: string) {
+		super(store, file, directory, { name: Events.RawMessageDelete, emitter: store.client.ws });
```
use discord-api-types event enum

kyranet replied:
Actually, that's a custom one, which processes the guild beforehand, named as the event's name is `'rawMessageDelete'`:

![image](https://user-images.githubusercontent.com/24852502/98996744-aa8f2600-2533-11eb-9c09-b83a6854942a.png)

vladfrangu replied:
Yet you listen to the ws manager.

vladfrangu commented on src/events/messages/rawMessageDeleteStarboard.ts:
```diff
@@ -1,28 +1,16 @@
 export default class extends Event {
 	public constructor(store: EventStore, file: string[], directory: string) {
-		super(store, file, directory, { name: DiscordEvents.MessageDelete, emitter: store.client.ws });
+		super(store, file, directory, { name: Events.RawMessageDelete, emitter: store.client.ws });
```
Ditto

vladfrangu commented on src/events/reactions/rawMessageReactionAdd.ts:
```diff
@@ -0,0 +1,42 @@
+				name: raw.emoji.name,
+				requireColons: raw.emoji.require_colons ?? null,
+				roles: raw.emoji.roles || null,
+				user: (raw.emoji.user && this.client.users.add(raw.emoji.user)) ?? { id: raw.user_id }
```
just a tip: if you don't want the user cached, add a false at the end

vladfrangu commented on src/lib/database/entities/StarboardEntity.ts:
```diff
@@ -248,41 +226,76 @@ export class StarboardEntity extends BaseEntity {
+	 * The text
+	 */
+	private getContent(language: Language) {
+		const url = `[${language.get(LanguageKeys.Misc.JumpTo)}](${this.#message.url})`;
```
Honestly, this pattern should be done in the translation string (so you end up with `language.get(JumpTo, { url })`

vladfrangu reviewed (changes requested):
I really cba to re-review the rest of the files, enjoy

kyranet committed:
fix: requested changes, part V

vladfrangu commented on src/commands/Moderation/Utilities/time.ts:
```diff
@@ -42,29 +44,33 @@ export default class extends SkyraCommand {
-		if (task)
-			throw await message.fetchLocale(LanguageKeys.Commands.Moderation.ModerationTimed, {
+		if (entry.appealType || entry.invalidated) {
+			throw await message.fetchLocale(LanguageKeys.Commands.Moderation.ModerationLogAppealed);
```
You already have the language instance?

vladfrangu commented on src/commands/Social/pay.ts:
```diff
@@ -64,14 +64,16 @@ export default class extends SkyraCommand {
+	private acceptPayment(message: GuildMessage, language: Language, user: User, money: number) {
 		this.client.emit(Events.MoneyPayment, message, message.author, user, money);
-		return message.sendMessage(language.get(LanguageKeys.Commands.Social.PayPromptAccept, { user: user.username, amount: money }));
+		return language.get(LanguageKeys.Commands.Social.PayPromptAccept, { user: user.username, amount: money });
```
This won't return to the user?

kyranet replied:
![image](https://user-images.githubusercontent.com/24852502/99068393-bcf97600-25ac-11eb-9a77-7e42e488c671.png)

vladfrangu commented on src/commands/Social/pay.ts:
```diff
@@ -64,14 +64,16 @@ export default class extends SkyraCommand {
-	private async denyPayment(message: GuildMessage, language: Language) {
-		return message.sendMessage(language.get(LanguageKeys.Commands.Social.PayPromptDeny));
+	private denyPayment(language: Language) {
+		return language.get(LanguageKeys.Commands.Social.PayPromptDeny);
```
Ditto

kyranet replied:
![Ditto](https://play.pokemonshowdown.com/sprites/ani/ditto.gif)

kyranet committed:
fix: requested changes, part VI

favna committed:
chore: migrations \o/

kyranet committed:
fix: use settings.fetchLanguage() over message.fetchLocale where possible

kyranet committed:
chore: more GuildMessage, optimized all serializers

kyranet committed:
chore: more refactors, allowed a lot more commands to run in DMs

favna committed:
fix: permission inhibitor works for dms

favna committed:
chore: allow more commands in DM

favna committed:
chore: formatting

favna committed:
fix: permission actual fix

favna committed:
feat: add dashonlyOnly configurable keys

favna committed:
fix: parse arrays in piece serializer

kyranet committed:
fix: resolved a couple of issues, handle arrays in dashboard correctly

kyranet committed:
style: minor refactor

kyranet committed:
fix: allow null in dashboard to reset

kyranet committed:
fix: assign default in SchemaKey

kyranet committed:
fix: prefix errors from settings endpoint with the key

depfu[bot] committed:
Update all Yarn dependencies (2020-11-15) (#1373)

Co-authored-by: depfu[bot] <23717796+depfu[bot]@users.noreply.github.com>

favna committed:
fix: remove array from configurable-key

kyranet committed:
fix: resolved 3 bugs

favna committed:
fix: suggest wasnt increasing ID

kyranet committed:
fix: resolved small error in commandError

kyranet committed:
fix: strictly check if the emoji is resolved

vladfrangu commented on src/commands/Admin/conf.ts:
```diff
@@ -17,88 +18,94 @@ import { KlasaMessage, SettingsFolder } from 'klasa';
 
-	public async init() {
-		initConfigurableSchema(this.client.gateways.get('guilds')!.schema);
+		return message.send(language.get(LanguageKeys.Commands.Admin.ConfUpdated, { key, response }), { allowedMentions: { users: [], roles: [] } });
```
Ok literally nitpick why tf is this one line and the one in set not??

vladfrangu commented on src/commands/Admin/roleset.ts:
```diff
@@ -36,110 +36,93 @@ import { KlasaMessage } from 'klasa';
-		}
+		const [created, roleSets, language] = await message.guild.writeSettings((settings) => {
+			const allRoleSets = settings[GuildSettings.Roles.UniqueRoleSets];
+			const roleSet = allRoleSets.find((set) => set.name === name);
```
You should probably use `some` instead, since you don't care of the value this returns, only if it exists

vladfrangu commented on src/commands/Admin/roleset.ts:
```diff
@@ -36,110 +36,93 @@ import { KlasaMessage } from 'klasa';
+			});
 
-			return { name, roles: finalRoleIDs };
+			return [false, settings[GuildSettings.Roles.UniqueRoleSets], settings.getLanguage()];
```
```suggestion
			return [false, allRoleSets, settings.getLanguage()];
```

vladfrangu commented on src/commands/Admin/roleset.ts:
```diff
@@ -36,110 +36,93 @@ import { KlasaMessage } from 'klasa';
-		return message.sendLocale(LanguageKeys.Commands.Admin.RolesetAdded, [
-			{
+		return message.send(
+			language.get(created ? LanguageKeys.Commands.Admin.RolesetCreated : LanguageKeys.Commands.Admin.RolesetAdded, {
```
Capitalization

vladfrangu commented on src/commands/Admin/roleset.ts:
```diff
@@ -36,110 +36,93 @@ import { KlasaMessage } from 'klasa';
-					message.language.get(LanguageKeys.Globals.And)
+				roles: language.list(
+					roleSets.map((role) => role.name),
+					language.get(LanguageKeys.Globals.And)
```
chore: make aliases in Language for `and`, `or`, etc.

favna replied:
That wouldn't be compliant with i18next spec that we're moving towards

vladfrangu commented on src/commands/Admin/roleset.ts:
```diff
@@ -36,110 +36,93 @@ import { KlasaMessage } from 'klasa';
 
 		const arrayIndex = allRolesets.findIndex((roleset) => roleset.name === name);
-		if (arrayIndex === -1) throw message.language.get(LanguageKeys.Commands.Admin.RolesetResetNotExists, { name });
+		if (arrayIndex === -1) throw language.get(LanguageKeys.Commands.Admin.RolesetResetNotExists, { name });
```
Ditto

vladfrangu commented on src/commands/Admin/roleset.ts:
```diff
@@ -36,110 +36,93 @@ import { KlasaMessage } from 'klasa';
 		});
-		return message.sendLocale(LanguageKeys.Commands.Admin.RolesetResetGroup, [{ name }]);
+
+		return message.send(language.get(LanguageKeys.Commands.Admin.RolesetResetGroup, { name }));
```
Diiiiito

vladfrangu commented on src/commands/Admin/roleset.ts:
```diff
@@ -150,26 +133,25 @@ export default class extends SkyraCommand {
 			// This will also remove any of the roles that user provided and were already in the set
-			const newroles = set.roles.map((id) => (roles.find((role) => role.id === id) ? null : id)).filter((id) => id);
+			const newroles = set.roles
+				.map((id) => (roles.find((role) => role.id === id) ? null : id)) //
```
Use .some

vladfrangu commented on src/commands/Admin/roleset.ts:
```diff
@@ -150,26 +133,25 @@ export default class extends SkyraCommand {
-			extraContext: { author: message.author.id }
-		});
+		await message.guild.writeSettings([[GuildSettings.Roles.UniqueRoleSets, newsets]]);
 		return message.sendLocale(LanguageKeys.Commands.Admin.RolesetUpdated, [{ name }]);
```
Ditto~

vladfrangu commented on src/commands/Developers/yarn.ts:
```diff
@@ -34,42 +33,45 @@ export default class extends SkyraCommand {
 			return await fetch<YarnPkg.PackageJson>(`https://registry.yarnpkg.com/${pkg}`, FetchResultTypes.JSON);
 		} catch {
-			throw message.language.get(LanguageKeys.Commands.Developers.YarnPackageNotFound, { pkg });
+			throw language.get(LanguageKeys.Commands.Developers.YarnPackageNotFound, { pkg });
```
In the future, use full variable names for translation keys (`package` not `pkg`)

kyranet replied:
`package` is a reserved variable name in JavaScript.

vladfrangu commented on src/commands/Games/coinflip.ts:
```diff
@@ -17,23 +17,44 @@ const enum CoinType {
+			if (!arg) return null;
+			const language = await message.fetchLanguage();
+			const lArg = arg.toLowerCase();
+			const face = language.get(LanguageKeys.Commands.Games.CoinFlipCoinnames).findIndex((coin) => coin.toLowerCase() === lArg);
```
Capitalization

vladfrangu commented on src/commands/Games/coinflip.ts:
```diff
@@ -17,23 +17,44 @@ const enum CoinType {
+			const language = await message.fetchLanguage();
+			const lArg = arg.toLowerCase();
+			const face = language.get(LanguageKeys.Commands.Games.CoinFlipCoinnames).findIndex((coin) => coin.toLowerCase() === lArg);
+			if (face === -1) throw language.get(LanguageKeys.Commands.Games.CoinFlipInvalidCoinname, { arg });
```
Ditto

vladfrangu commented on src/commands/Management/Configuration/manageCommandChannel.ts:
```diff
@@ -23,101 +24,103 @@ export default class extends SkyraCommand {
 				if (!command.disabled && command.permissionLevel < 9) return command;
 			}
-			throw msg.language.get(LanguageKeys.Commands.Management.ManageCommandChannelRequiredCommand);
+			throw msg.fetchLocale(LanguageKeys.Commands.Management.ManageCommandChannelRequiredCommand);
```
throw await

Move to decorator

vladfrangu commented on src/commands/Management/Configuration/manageCommandChannel.ts:
```diff
@@ -23,101 +24,103 @@ export default class extends SkyraCommand {
+			} else {
+				const disabledCommandChannel = disabledCommandsChannels[indexOfChannel];
+				if (disabledCommandChannel.commands.includes(command.name))
+					throw language.get(LanguageKeys.Commands.Management.ManageCommandChannelAddAlreadyset);
```
Capitalization

vladfrangu commented on src/commands/Management/Configuration/manageCommandChannel.ts:
```diff
@@ -23,101 +24,103 @@ export default class extends SkyraCommand {
+				if (disabledCommandChannel.commands.includes(command.name))
+					throw language.get(LanguageKeys.Commands.Management.ManageCommandChannelAddAlreadyset);
+
+				settings[GuildSettings.DisabledCommandChannels][indexOfChannel].commands.concat(command.name);
```
This won't do anything; concat returns a new array. Did you mean `push`?
```suggestion
				settings[GuildSettings.DisabledCommandChannels][indexOfChannel].commands.push(command.name);
```

vladfrangu commented on src/commands/Management/Configuration/manageCommandChannel.ts:
```diff
@@ -23,101 +24,103 @@ export default class extends SkyraCommand {
+			const language = settings.getLanguage();
+
+			if (indexOfChannel === -1) {
+				throw language.get(LanguageKeys.Commands.Management.ManageCommandChannelRemoveNotset, { channel: channel.toString() });
```
Capitalization

favna committed:
refactor: fix and improve suggestion ID counting

favna committed:
chore: typings tweak

vladfrangu reviewed (changes requested):
I really cba to go through everything again...so, have these, make testers test it, otherwise probably LGTM..

kyranet committed:
fix: requested changes, part VII

kyranet committed:
refactor: make suggest faster

kyranet committed:
fix: screw i18n, gonna hardcode the guild's name

vladfrangu reviewed (dismissed):
Get testers first, then deploy to prod

favna committed:
chore: change klasa dep to settings branch

vladfrangu reviewed (approved):
I'm not checking nearly a symmetrical 10k diff PR, just merge it
